// claude-to-im/src/lib/bridge/bridge-manager.ts
/**
 * Bridge Manager 閳?singleton orchestrator for the multi-IM bridge system.
 *
 * Manages adapter lifecycles, routes inbound messages through the
 * conversation engine, and coordinates permission handling.
 *
 * Uses globalThis to survive Next.js HMR in development.
 */

import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

import type { BridgeStatus, InboundMessage, OutboundMessage, StreamingPreviewState, ToolCallInfo } from './types.js';
import { createAdapter, getRegisteredTypes } from './channel-adapter.js';
import type { BaseChannelAdapter } from './channel-adapter.js';
import type { BridgeMessage, BridgeSession } from './host.js';
// Side-effect import: triggers self-registration of all adapter factories
import './adapters/index.js';
import * as router from './channel-router.js';
import * as engine from './conversation-engine.js';
import * as broker from './permission-broker.js';
import { deliver, deliverRendered } from './delivery-layer.js';
import { markdownToTelegramChunks } from './markdown/telegram.js';
import { markdownToDiscordChunks } from './markdown/discord.js';
import { getBridgeContext } from './context.js';
import { escapeHtml } from './adapters/telegram-utils.js';
import {
  validateWorkingDirectory,
  validateSessionId,
  isDangerousInput,
  sanitizeInput,
  validateMode,
} from './security/validators.js';

const GLOBAL_KEY = '__bridge_manager__';

// 閳光偓閳光偓 Streaming preview helpers 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓

/** Generate a non-zero random 31-bit integer for use as draft_id. */
function generateDraftId(): number {
  return (Math.floor(Math.random() * 0x7FFFFFFE) + 1); // 1 .. 2^31-1
}

interface StreamConfig {
  intervalMs: number;
  minDeltaChars: number;
  maxChars: number;
}

/** Default stream config per channel type. */
const STREAM_DEFAULTS: Record<string, StreamConfig> = {
  telegram: { intervalMs: 700, minDeltaChars: 20, maxChars: 3900 },
  discord: { intervalMs: 1500, minDeltaChars: 40, maxChars: 1900 },
};

function getStreamConfig(channelType = 'telegram'): StreamConfig {
  const { store } = getBridgeContext();
  const defaults = STREAM_DEFAULTS[channelType] || STREAM_DEFAULTS.telegram;
  const prefix = `bridge_${channelType}_stream_`;
  const intervalMs = parseInt(store.getSetting(`${prefix}interval_ms`) || '', 10) || defaults.intervalMs;
  const minDeltaChars = parseInt(store.getSetting(`${prefix}min_delta_chars`) || '', 10) || defaults.minDeltaChars;
  const maxChars = parseInt(store.getSetting(`${prefix}max_chars`) || '', 10) || defaults.maxChars;
  return { intervalMs, minDeltaChars, maxChars };
}

/**
 * Check if a message looks like a numeric permission shortcut (1/2/3) for
 * feishu/qq channels WITH at least one pending permission in that chat.
 *
 * This is used by the adapter loop to route these messages to the inline
 * (non-session-locked) path, avoiding deadlock: the session is blocked
 * waiting for the permission to be resolved, so putting "1" behind the
 * session lock would deadlock.
 */
function isNumericPermissionShortcut(channelType: string, rawText: string, chatId: string): boolean {
  if (channelType !== 'feishu' && channelType !== 'qq' && channelType !== 'weixin') return false;
  const normalized = rawText.normalize('NFKC').replace(/[\u200B-\u200D\uFEFF]/g, '').trim();
  if (!/^[123]$/.test(normalized)) return false;
  const { store } = getBridgeContext();
  const pending = store.listPendingPermissionLinksByChat(chatId);
  return pending.length > 0; // any pending 閳?route to inline path
}

/** Fire-and-forget: send a preview draft. Only degrades on permanent failure. */
function flushPreview(
  adapter: BaseChannelAdapter,
  state: StreamingPreviewState,
  config: StreamConfig,
): void {
  if (state.degraded || !adapter.sendPreview) return;

  const text = state.pendingText.length > config.maxChars
    ? state.pendingText.slice(0, config.maxChars) + '...'
    : state.pendingText;

  state.lastSentText = text;
  state.lastSentAt = Date.now();

  adapter.sendPreview(state.chatId, text, state.draftId).then(result => {
    if (result === 'degrade') state.degraded = true;
    // 'skip' 閳?transient failure, next flush will retry naturally
  }).catch(() => {
    // Network error 閳?transient, don't degrade
  });
}

// 閳光偓閳光偓 Channel-aware rendering dispatch 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓

import type { ChannelAddress, ChannelBinding, SendResult } from './types.js';

/**
 * Render response text and deliver via the appropriate channel format.
 * Telegram: Markdown 閳?HTML chunks via deliverRendered.
 * Other channels: plain text via deliver (no HTML).
 */
async function deliverResponse(
  adapter: BaseChannelAdapter,
  address: ChannelAddress,
  responseText: string,
  sessionId: string,
  replyToMessageId?: string,
): Promise<SendResult> {
  if (adapter.channelType === 'telegram') {
    const chunks = markdownToTelegramChunks(responseText, 4096);
    if (chunks.length > 0) {
      return deliverRendered(adapter, address, chunks, { sessionId, replyToMessageId });
    }
    return { ok: true };
  }
  if (adapter.channelType === 'discord') {
    // Discord: native markdown, chunk at 2000 chars with fence repair
    const chunks = markdownToDiscordChunks(responseText, 2000);
    for (let i = 0; i < chunks.length; i++) {
      const result = await deliver(adapter, {
        address,
        text: chunks[i].text,
        parseMode: 'Markdown',
        replyToMessageId,
      }, { sessionId });
      if (!result.ok) return result;
    }
    return { ok: true };
  }
  if (adapter.channelType === 'feishu') {
    // Feishu: pass markdown through for adapter to format as post/card
    return deliver(adapter, {
      address,
      text: responseText,
      parseMode: 'Markdown',
      replyToMessageId,
    }, { sessionId });
  }
  // Generic fallback: deliver as plain text (deliver() handles chunking internally)
  return deliver(adapter, {
    address,
    text: responseText,
    parseMode: 'plain',
    replyToMessageId,
  }, { sessionId });
}

type NamedBridgeSession = BridgeSession & {
  name?: string;
  sdk_session_id?: string;
  source?: 'bridge' | 'codex';
  created_at?: string;
  updated_at?: string;
  native_log_path?: string;
};

type SessionListingStore = ReturnType<typeof getBridgeContext>['store'] & {
  listSessions?: () => NamedBridgeSession[];
  updateSessionName?: (sessionId: string, name: string) => void;
  updateSessionMetadata?: (
    sessionId: string,
    updates: Partial<Pick<NamedBridgeSession, 'name' | 'working_directory' | 'model' | 'created_at' | 'updated_at'>>,
  ) => void;
  archiveSession?: (sessionId: string) => void;
  updateSdkSessionId?: (sessionId: string, sdkSessionId: string) => void;
};

interface ProjectGroup {
  path: string;
  label: string;
  sessions: NamedBridgeSession[];
  current: boolean;
}

interface WorkspaceGroup {
  path: string;
  label: string;
  projectCount: number;
  sessionCount: number;
  current: boolean;
}

interface SessionContextPreview {
  session: NamedBridgeSession;
  messages: BridgeMessage[];
  page: number;
  pageSize: number;
  totalMessages: number;
  totalPages: number;
  startIndex: number;
  endIndex: number;
}

const CONTEXT_PAGE_SIZE = 10;
const PROJECT_SESSIONS_PAGE_SIZE = 5;

interface FeishuSessionCardItem {
  sessionId: string;
  title: string;
  subtitle: string;
  timestamps: string;
  current: boolean;
  messageCount: number;
}

type DockAwareBinding = ChannelBinding & {
  openSessionIds?: string[];
  sessionSeenCounts?: Record<string, number>;
};

interface SessionDockItem {
  sessionId: string;
  projectLabel: string;
  title: string;
  current: boolean;
  running: boolean;
  unreadCount: number;
  messageCount: number;
}

const SESSION_DOCK_LIMIT = 6;
type ProjectListScope = 'workspace' | 'global';
const CODEX_DISCOVERY_CACHE_MS = 30_000;

type CodexDiscoveryCache = {
  home: string;
  indexMtimeMs: number;
  sessionsMtimeMs: number;
  loadedAt: number;
  sessions: NamedBridgeSession[];
};

type WorkspaceDiscoveryCache = {
  home: string;
  globalStateMtimeMs: number;
  roots: string[];
  labels: Record<string, string>;
  active: Set<string>;
};

let codexDiscoveryCache: CodexDiscoveryCache | null = null;
const nativeSessionMessageCache = new Map<string, { mtimeMs: number; messages: BridgeMessage[] }>();
let workspaceDiscoveryCache: WorkspaceDiscoveryCache | null = null;

function getSessionListingStore(): SessionListingStore {
  return getBridgeContext().store as SessionListingStore;
}

function listBridgeSessions(): NamedBridgeSession[] {
  const store = getSessionListingStore();
  if (typeof store.listSessions === 'function') {
    const sessions = store.listSessions().map((session) => syncBridgeSessionFromCodex({ ...session, source: 'bridge' as const }));
    const boundSessionIds = new Set(store.listChannelBindings().map((binding) => binding.codepilotSessionId));
    const deduped = new Map<string, NamedBridgeSession>();

    for (const session of sessions) {
      const sdkSessionId = getStoredSdkSessionId(session);
      const dedupeKey = sdkSessionId || session.id;
      const existing = deduped.get(dedupeKey);
      if (!existing) {
        deduped.set(dedupeKey, session);
        continue;
      }

      const existingBound = boundSessionIds.has(existing.id);
      const nextBound = boundSessionIds.has(session.id);
      if (existingBound !== nextBound) {
        deduped.set(dedupeKey, nextBound ? session : existing);
        continue;
      }

      const sortDelta = getSessionSortTime(session) - getSessionSortTime(existing);
      if (sortDelta !== 0) {
        deduped.set(dedupeKey, sortDelta > 0 ? session : existing);
        continue;
      }

      const messageDelta = getSessionMessageCount(session.id) - getSessionMessageCount(existing.id);
      if (messageDelta !== 0) {
        deduped.set(dedupeKey, messageDelta > 0 ? session : existing);
        continue;
      }

      if (session.id.localeCompare(existing.id, 'en') > 0) {
        deduped.set(dedupeKey, session);
      }
    }

    return Array.from(deduped.values());
  }
  return [];
}

function getCodexSessionById(sessionId: string | null | undefined): NamedBridgeSession | null {
  if (!sessionId) return null;
  return discoverCodexSessions().find((session) => session.id === sessionId) ?? null;
}

function syncBridgeSessionFromCodex(session: NamedBridgeSession): NamedBridgeSession {
  const sdkSessionId = getStoredSdkSessionId(session);
  if (!sdkSessionId) return session;

  const discovered = getCodexSessionById(sdkSessionId);
  if (!discovered) return session;

  const merged: NamedBridgeSession = {
    ...session,
    name: discovered.name || session.name,
    working_directory: discovered.working_directory || session.working_directory,
    model: discovered.model || session.model,
    created_at: discovered.created_at || session.created_at,
    updated_at: discovered.updated_at || session.updated_at,
  };

  if (
    merged.name !== session.name
    || merged.working_directory !== session.working_directory
    || merged.model !== session.model
    || merged.created_at !== session.created_at
    || merged.updated_at !== session.updated_at
  ) {
    getSessionListingStore().updateSessionMetadata?.(session.id, {
      name: merged.name,
      working_directory: merged.working_directory,
      model: merged.model,
      created_at: merged.created_at,
      updated_at: merged.updated_at,
    });
  }

  return merged;
}

function getCodexHome(): string {
  const configuredHome = process.env.CODEX_HOME?.trim();
  return configuredHome || path.join(os.homedir(), '.codex');
}

function safeStatMtimeMs(targetPath: string): number {
  try {
    return fs.statSync(targetPath).mtimeMs;
  } catch {
    return 0;
  }
}

function listSessionLogFiles(rootDir: string, out: string[] = []): string[] {
  let entries: fs.Dirent[];
  try {
    entries = fs.readdirSync(rootDir, { withFileTypes: true });
  } catch {
    return out;
  }
  for (const entry of entries) {
    const entryPath = path.join(rootDir, entry.name);
    if (entry.isDirectory()) {
      listSessionLogFiles(entryPath, out);
    } else if (entry.isFile() && entry.name.endsWith('.jsonl')) {
      out.push(entryPath);
    }
  }
  return out;
}

function readFileHead(filePath: string, maxBytes = 64 * 1024): string {
  let fd: number | null = null;
  try {
    fd = fs.openSync(filePath, 'r');
    const buffer = Buffer.alloc(maxBytes);
    const bytesRead = fs.readSync(fd, buffer, 0, maxBytes, 0);
    return buffer.toString('utf8', 0, bytesRead);
  } catch {
    return '';
  } finally {
    if (fd != null) {
      try { fs.closeSync(fd); } catch { /* best effort */ }
    }
  }
}

function getStoredSdkSessionId(session: NamedBridgeSession | null | undefined): string | null {
  const value = session?.sdk_session_id;
  return typeof value === 'string' && value.trim() ? value : null;
}

function getSessionSortTime(session: NamedBridgeSession): number {
  const candidates = [
    session.updated_at,
    (session as unknown as { updatedAt?: string }).updatedAt,
    session.created_at,
    (session as unknown as { createdAt?: string }).createdAt,
  ];
  for (const candidate of candidates) {
    if (typeof candidate !== 'string') continue;
    const timestamp = Date.parse(candidate);
    if (!Number.isNaN(timestamp)) return timestamp;
  }
  return 0;
}

function getSessionCreatedAt(session: NamedBridgeSession): string | null {
  const candidates = [
    session.created_at,
    (session as unknown as { createdAt?: string }).createdAt,
  ];
  for (const candidate of candidates) {
    if (typeof candidate === 'string' && candidate.trim()) return candidate;
  }
  return null;
}

function getSessionUpdatedAt(session: NamedBridgeSession): string | null {
  const candidates = [
    session.updated_at,
    (session as unknown as { updatedAt?: string }).updatedAt,
    session.created_at,
    (session as unknown as { createdAt?: string }).createdAt,
  ];
  for (const candidate of candidates) {
    if (typeof candidate === 'string' && candidate.trim()) return candidate;
  }
  return null;
}

function formatSessionTimestamp(value: string | null): string {
  if (!value) return '—';
  const date = new Date(value);
  if (Number.isNaN(date.getTime())) return '—';
  const year = String(date.getFullYear());
  const month = String(date.getMonth() + 1).padStart(2, '0');
  const day = String(date.getDate()).padStart(2, '0');
  const hours = String(date.getHours()).padStart(2, '0');
  const minutes = String(date.getMinutes()).padStart(2, '0');
  return `${year}-${month}-${day} ${hours}:${minutes}`;
}

function buildSessionTimestampSummary(session: NamedBridgeSession): string {
  return `Created ${formatSessionTimestamp(getSessionCreatedAt(session))} · Updated ${formatSessionTimestamp(getSessionUpdatedAt(session))}`;
}

function discoverCodexSessions(): NamedBridgeSession[] {
  const codexHome = getCodexHome();
  const indexPath = path.join(codexHome, 'session_index.jsonl');
  const sessionsDir = path.join(codexHome, 'sessions');
  const indexMtimeMs = safeStatMtimeMs(indexPath);
  const sessionsMtimeMs = safeStatMtimeMs(sessionsDir);
  const nowMs = Date.now();

  if (
    codexDiscoveryCache
    && codexDiscoveryCache.home === codexHome
    && codexDiscoveryCache.indexMtimeMs === indexMtimeMs
    && codexDiscoveryCache.sessionsMtimeMs === sessionsMtimeMs
    && nowMs - codexDiscoveryCache.loadedAt < CODEX_DISCOVERY_CACHE_MS
  ) {
    return codexDiscoveryCache.sessions;
  }

  const indexEntries = new Map<string, { title?: string; updatedAt?: string }>();
  try {
    const indexText = fs.readFileSync(indexPath, 'utf8');
    for (const line of indexText.split(/\r?\n/)) {
      if (!line.trim()) continue;
      try {
        const item = JSON.parse(line) as { id?: unknown; thread_name?: unknown; updated_at?: unknown };
        if (typeof item.id !== 'string' || !validateSessionId(item.id)) continue;
        indexEntries.set(item.id, {
          title: typeof item.thread_name === 'string' ? item.thread_name : undefined,
          updatedAt: typeof item.updated_at === 'string' ? item.updated_at : undefined,
        });
      } catch {
        // Ignore corrupt jsonl rows; Codex can append partial rows while running.
      }
    }
  } catch {
    // No Codex index is fine; bridge-only navigation still works.
  }

  const sessions = new Map<string, NamedBridgeSession>();
  for (const filePath of listSessionLogFiles(sessionsDir)) {
    const head = readFileHead(filePath);
    const metaLine = head.split(/\r?\n/, 16).find((line) => line.includes('"type":"session_meta"'));
    if (!metaLine) continue;
    try {
      const event = JSON.parse(metaLine) as {
        timestamp?: unknown;
        payload?: { id?: unknown; cwd?: unknown; timestamp?: unknown };
      };
      const id = event.payload?.id;
      const cwd = event.payload?.cwd;
      if (typeof id !== 'string' || !validateSessionId(id) || typeof cwd !== 'string' || !cwd.trim()) {
        continue;
      }
      const indexed = indexEntries.get(id);
      sessions.set(id, {
        id,
        name: indexed?.title || `Codex ${getShortSessionId(id)}`,
        working_directory: normalizeProjectPath(cwd),
        model: '',
        system_prompt: undefined,
        source: 'codex',
        created_at: (typeof event.payload?.timestamp === 'string' ? event.payload.timestamp : undefined)
          || (typeof event.timestamp === 'string' ? event.timestamp : undefined),
        updated_at: indexed?.updatedAt || (typeof event.payload?.timestamp === 'string' ? event.payload.timestamp : undefined)
          || (typeof event.timestamp === 'string' ? event.timestamp : undefined),
        native_log_path: filePath,
      } as NamedBridgeSession);
    } catch {
      // Ignore individual broken session files.
    }
  }

  const discovered = Array.from(sessions.values()).sort((left, right) => getSessionSortTime(right) - getSessionSortTime(left));
  codexDiscoveryCache = {
    home: codexHome,
    indexMtimeMs,
    sessionsMtimeMs,
    loadedAt: nowMs,
    sessions: discovered,
  };
  return discovered;
}

function listKnownSessions(): NamedBridgeSession[] {
  const bridgeSessions = listBridgeSessions();
  const knownIds = new Set(bridgeSessions.map((session) => session.id));
  const knownSdkIds = new Set(
    bridgeSessions
      .map((session) => getStoredSdkSessionId(session))
      .filter((sessionId): sessionId is string => Boolean(sessionId)),
  );
  const codexSessions = discoverCodexSessions()
    .filter((session) => !knownIds.has(session.id) && !knownSdkIds.has(session.id));
  return [...bridgeSessions, ...codexSessions];
}

function clearCodexDiscoveryCache(): void {
  codexDiscoveryCache = null;
  nativeSessionMessageCache.clear();
  workspaceDiscoveryCache = null;
}

function normalizeProjectPath(rawPath: string | null | undefined): string {
  const trimmed = rawPath?.trim();
  if (!trimmed) return '~';
  const withoutDevicePrefix = trimmed.startsWith('\\\\?\\') ? trimmed.slice(4) : trimmed;
  const normalized = withoutDevicePrefix.replace(/\//g, '\\').replace(/[\\]+$/, '');
  if (/^[a-zA-Z]:\\/.test(normalized)) {
    return `${normalized[0]!.toUpperCase()}${normalized.slice(1)}`;
  }
  return normalized;
}

function getProjectLabel(projectPath: string): string {
  const normalized = projectPath.replace(/[\\/]+$/, '');
  const base = path.basename(normalized);
  return base || normalized || '~';
}

function getShortSessionId(sessionId: string): string {
  return sessionId.slice(0, 8);
}

function getSessionTitle(session: NamedBridgeSession): string {
  const explicitName = session.name?.trim();
  if (explicitName && !explicitName.startsWith('Bridge:')) {
    return explicitName;
  }
  return getShortSessionId(session.id);
}

function formatSessionCount(count: number): string {
  return `${count} session${count === 1 ? '' : 's'}`;
}

function formatProjectCount(count: number): string {
  return `${count} project${count === 1 ? '' : 's'}`;
}

function formatMessageCount(count: number): string {
  return `${count} msg${count === 1 ? '' : 's'}`;
}

function getProjectPathHint(projectPath: string): string {
  const normalized = projectPath.replace(/[\\/]+$/, '');
  const parent = path.dirname(normalized);
  if (!parent || parent === '.' || parent === normalized) {
    return normalized;
  }
  return parent.length > 28 ? `...${parent.slice(-25)}` : parent;
}

function discoverCodexWorkspaceState(): WorkspaceDiscoveryCache {
  const codexHome = getCodexHome();
  const globalStatePath = path.join(codexHome, '.codex-global-state.json');
  const globalStateMtimeMs = safeStatMtimeMs(globalStatePath);

  if (
    workspaceDiscoveryCache
    && workspaceDiscoveryCache.home === codexHome
    && workspaceDiscoveryCache.globalStateMtimeMs === globalStateMtimeMs
  ) {
    return workspaceDiscoveryCache;
  }

  const roots = new Set<string>();
  const labels: Record<string, string> = {};
  const active = new Set<string>();

  try {
    const raw = fs.readFileSync(globalStatePath, 'utf8');
    const state = JSON.parse(raw) as {
      'electron-saved-workspace-roots'?: unknown;
      'active-workspace-roots'?: unknown;
      'electron-workspace-root-labels'?: unknown;
    };

    for (const item of Array.isArray(state['electron-saved-workspace-roots']) ? state['electron-saved-workspace-roots'] : []) {
      if (typeof item !== 'string') continue;
      const normalized = normalizeProjectPath(item);
      if (normalized) roots.add(normalized);
    }

    for (const item of Array.isArray(state['active-workspace-roots']) ? state['active-workspace-roots'] : []) {
      if (typeof item !== 'string') continue;
      const normalized = normalizeProjectPath(item);
      if (!normalized) continue;
      roots.add(normalized);
      active.add(normalized);
    }

    const rawLabels = state['electron-workspace-root-labels'];
    if (rawLabels && typeof rawLabels === 'object' && !Array.isArray(rawLabels)) {
      for (const [key, value] of Object.entries(rawLabels as Record<string, unknown>)) {
        const normalizedKey = normalizeProjectPath(key);
        if (typeof value === 'string' && value.trim()) {
          labels[normalizedKey] = value.trim();
        }
      }
    }
  } catch {
    // Missing global state is okay; we'll fall back to per-project workspaces.
  }

  workspaceDiscoveryCache = {
    home: codexHome,
    globalStateMtimeMs,
    roots: Array.from(roots).sort((left, right) => right.length - left.length),
    labels,
    active,
  };
  return workspaceDiscoveryCache;
}

function getWorkspaceLabel(workspaceRoot: string): string {
  const state = discoverCodexWorkspaceState();
  return state.labels[workspaceRoot] || getProjectLabel(workspaceRoot);
}

function getContainingWorkspaceRoot(projectPath: string): string | null {
  const normalized = normalizeProjectPath(projectPath);
  const state = discoverCodexWorkspaceState();
  const lowerPath = normalized.toLowerCase();
  for (const workspaceRoot of state.roots) {
    const lowerRoot = workspaceRoot.toLowerCase();
    if (lowerPath === lowerRoot || lowerPath.startsWith(`${lowerRoot}\\`)) {
      return workspaceRoot;
    }
  }
  return null;
}

function isKnownWorkspaceRoot(workspacePath: string): boolean {
  const normalized = normalizeProjectPath(workspacePath);
  const state = discoverCodexWorkspaceState();
  return state.roots.some((root) => root.toLowerCase() === normalized.toLowerCase());
}

function getProjectNavigationScope(projectPath: string): ProjectListScope {
  return getContainingWorkspaceRoot(projectPath) ? 'workspace' : 'global';
}

function isInWorkspace(projectPath: string, workspaceRoot: string): boolean {
  const normalizedProject = normalizeProjectPath(projectPath).toLowerCase();
  const normalizedWorkspace = normalizeProjectPath(workspaceRoot).toLowerCase();
  return normalizedProject === normalizedWorkspace || normalizedProject.startsWith(`${normalizedWorkspace}\\`);
}

function getCompactProjectPath(projectPath: string): string {
  const normalized = projectPath.replace(/\//g, '\\').replace(/[\\]+$/, '');
  const parts = normalized.split(/\\+/).filter(Boolean);
  if (parts.length <= 2) return normalized;
  const tail = parts.slice(-2).join('\\');
  const prefix = /^[a-zA-Z]:$/.test(parts[0] || '') ? `${parts[0]}\\...\\` : '...\\';
  return `${prefix}${tail}`;
}

function getCurrentSessionLabel(binding: ChannelBinding): string {
  const currentSession = getCurrentSession(binding);
  const label = currentSession ? getSessionTitle(currentSession) : getShortSessionId(binding.codepilotSessionId);
  const singleLine = label.replace(/\s+/g, ' ').trim();
  if (!singleLine) {
    return getShortSessionId(binding.codepilotSessionId);
  }
  const compact = singleLine.startsWith('Session ') ? singleLine.slice('Session '.length) : singleLine;
  return compact.length > 24 ? `${compact.slice(0, 21)}...` : compact;
}

function getCurrentSession(binding: ChannelBinding): NamedBridgeSession | null {
  return getSessionDisplayRecord(binding.codepilotSessionId);
}

function getSessionDisplayRecord(sessionId: string): NamedBridgeSession | null {
  const store = getSessionListingStore();
  const direct = store.getSession(sessionId) as NamedBridgeSession | null;
  if (direct) {
    return syncBridgeSessionFromCodex({ ...direct, source: 'bridge' });
  }
  const discovered = getCodexSessionById(sessionId);
  return discovered ? { ...discovered, source: 'codex' } : null;
}

function syncBindingToCurrentSession(binding: ChannelBinding): ChannelBinding {
  const store = getSessionListingStore();
  const currentSession = getSessionDisplayRecord(binding.codepilotSessionId);
  if (!currentSession) return binding;

  const nextWorkingDirectory = normalizeProjectPath(currentSession.working_directory || binding.workingDirectory);
  const nextModel = currentSession.model || binding.model || '';
  const nextSdkSessionId = getStoredSdkSessionId(currentSession) || binding.sdkSessionId || '';
  const updates: Partial<ChannelBinding> = {};

  if (nextWorkingDirectory && normalizeProjectPath(binding.workingDirectory) !== nextWorkingDirectory) {
    updates.workingDirectory = nextWorkingDirectory;
  }
  if (nextModel !== binding.model) {
    updates.model = nextModel;
  }
  if (nextSdkSessionId !== (binding.sdkSessionId || '')) {
    updates.sdkSessionId = nextSdkSessionId;
  }

  if (Object.keys(updates).length === 0) {
    return binding;
  }

  store.updateChannelBinding(binding.id, updates);
  return store.getChannelBinding(binding.channelType, binding.chatId) ?? { ...binding, ...updates };
}

function buildCurrentSessionPanel(
  binding: ChannelBinding,
  options?: {
    title?: string;
    showProject?: boolean;
    showTimestamps?: boolean;
    buttons?: Array<{ label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> }>;
  },
): Record<string, unknown> {
  const currentSession = getCurrentSession(binding);
  const fallbackProjectPath = normalizeProjectPath(binding.workingDirectory);
  const projectPath = normalizeProjectPath(currentSession?.working_directory || fallbackProjectPath);
  const session = currentSession ?? {
    id: binding.codepilotSessionId,
    name: getShortSessionId(binding.codepilotSessionId),
    working_directory: projectPath,
    model: binding.model || '',
  } as NamedBridgeSession;
  const meta = `${getShortSessionId(session.id)} · ${formatMessageCount(getSessionMessageCount(session.id))} · ${session.model || binding.model || 'default'}`;
  const lines = [
    `**${escapeCardText(options?.title || 'Current Session')}**`,
    `**${escapeCardText(getSessionTitle(session))}**`,
    `\`${escapeCardText(meta)}\``,
    options?.showProject === false
      ? null
      : `${escapeCardText(getProjectLabel(projectPath))} · \`${escapeCardText(getCompactProjectPath(projectPath))}\``,
    options?.showTimestamps === false ? null : escapeCardText(buildSessionTimestampSummary(session)),
  ].filter((line): line is string => Boolean(line));

  const elements: Array<Record<string, unknown>> = [
    {
      tag: 'markdown',
      content: lines.join('\n'),
      text_size: 'normal',
    },
  ];

  if (options?.buttons?.length) {
    elements.push(createInlineCardButtonSet(options.buttons));
  }

  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements,
      },
    ],
  };
}

function getShortDockTitle(session: NamedBridgeSession): string {
  const title = getSessionTitle(session).replace(/\s+/g, ' ').trim() || getShortSessionId(session.id);
  const compact = title.startsWith('Session ') ? title.slice('Session '.length) : title;
  return compact.length > 10 ? `${compact.slice(0, 8)}…` : compact;
}

function getShortDockProjectLabel(projectPath: string): string {
  const label = getProjectLabel(projectPath).replace(/\s+/g, ' ').trim() || '~';
  return label.length > 9 ? `${label.slice(0, 7)}…` : label;
}

function getDockSeenCounts(binding: ChannelBinding): Record<string, number> {
  const raw = (binding as DockAwareBinding).sessionSeenCounts;
  return raw && typeof raw === 'object' && !Array.isArray(raw) ? { ...raw } : {};
}

function getOpenSessionIds(binding: ChannelBinding): string[] {
  const store = getSessionListingStore();
  const rawIds = Array.isArray((binding as DockAwareBinding).openSessionIds)
    ? (binding as DockAwareBinding).openSessionIds!
    : [];
  const ids = [binding.codepilotSessionId, ...rawIds]
    .filter((id): id is string => typeof id === 'string' && validateSessionId(id));
  const uniqueIds = Array.from(new Set(ids));
  return uniqueIds
    .filter((sessionId) => {
      const session = store.getSession(sessionId) as NamedBridgeSession | null;
      return Boolean(session && !session.archived_at);
    })
    .slice(0, SESSION_DOCK_LIMIT);
}

function persistSessionDock(
  binding: ChannelBinding,
  openSessionIds: string[],
  seenCounts = getDockSeenCounts(binding),
): ChannelBinding {
  const store = getSessionListingStore();
  const sanitizedIds = Array.from(new Set(openSessionIds.filter((id) => validateSessionId(id))))
    .slice(0, SESSION_DOCK_LIMIT);
  store.updateChannelBinding(binding.id, {
    openSessionIds: sanitizedIds,
    sessionSeenCounts: seenCounts,
  } as Partial<ChannelBinding>);
  return store.getChannelBinding(binding.channelType, binding.chatId) ?? binding;
}

function addSessionToDock(binding: ChannelBinding, sessionId: string, markSeen = false): ChannelBinding {
  const store = getSessionListingStore();
  if (!validateSessionId(sessionId) || !store.getSession(sessionId)) return binding;

  const openSessionIds = [sessionId, ...getOpenSessionIds(binding).filter((id) => id !== sessionId)]
    .slice(0, SESSION_DOCK_LIMIT);
  const seenCounts = getDockSeenCounts(binding);
  if (markSeen) {
    seenCounts[sessionId] = getSessionMessageCount(sessionId);
  }
  return persistSessionDock(binding, openSessionIds, seenCounts);
}

function removeSessionFromDock(binding: ChannelBinding, sessionId: string): ChannelBinding {
  const openSessionIds = getOpenSessionIds(binding).filter((id) => id !== sessionId);
  const seenCounts = getDockSeenCounts(binding);
  delete seenCounts[sessionId];
  return persistSessionDock(binding, openSessionIds, seenCounts);
}

function markDockSessionSeen(binding: ChannelBinding, sessionId: string): ChannelBinding {
  if (!validateSessionId(sessionId)) return binding;
  const seenCounts = getDockSeenCounts(binding);
  seenCounts[sessionId] = getSessionMessageCount(sessionId);
  return persistSessionDock(binding, getOpenSessionIds(binding), seenCounts);
}

function getSessionDockItems(binding: ChannelBinding): SessionDockItem[] {
  const effectiveBinding = syncBindingToCurrentSession(binding);
  const seenCounts = getDockSeenCounts(effectiveBinding);
  const activeTasks = getState().activeTasks;
  return getOpenSessionIds(effectiveBinding)
    .map((sessionId) => {
      const session = getSessionDisplayRecord(sessionId);
      if (!session) return null;
      const messageCount = getSessionMessageCount(sessionId);
      const seenCount = seenCounts[sessionId] ?? (sessionId === effectiveBinding.codepilotSessionId ? messageCount : 0);
      return {
        sessionId,
        projectLabel: getShortDockProjectLabel(session.working_directory),
        title: getShortDockTitle(session),
        current: sessionId === effectiveBinding.codepilotSessionId,
        running: activeTasks.has(sessionId),
        unreadCount: Math.max(0, messageCount - seenCount),
        messageCount,
      };
    })
    .filter((item): item is SessionDockItem => Boolean(item));
}

function getDockIndicator(item: SessionDockItem): string {
  if (item.running) return '◉';
  if (item.unreadCount > 0 && !item.current) return '●';
  if (item.current) return '●';
  return '○';
}

function summarizePlainContextText(content: string, maxLength = 160): string {
  const singleLine = content.replace(/\s+/g, ' ').trim();
  if (!singleLine) return '(empty)';
  return singleLine.length > maxLength ? `${singleLine.slice(0, maxLength - 3)}...` : singleLine;
}

function summarizeStructuredContextText(content: string, maxLength = 160): string | null {
  const trimmed = content.trim();
  if (!trimmed || (!trimmed.startsWith('[') && !trimmed.startsWith('{'))) return null;

  try {
    const parsed = JSON.parse(trimmed);
    const blocks = Array.isArray(parsed) ? parsed : [parsed];
    if (!blocks.some((block) => block && typeof block === 'object' && typeof block.type === 'string')) {
      return null;
    }

    const textParts: string[] = [];
    const toolNames: string[] = [];

    for (const block of blocks) {
      if (!block || typeof block !== 'object') continue;
      const record = block as Record<string, unknown>;
      const type = typeof record.type === 'string' ? record.type : '';
      if (type === 'text' && typeof record.text === 'string') {
        const text = summarizePlainContextText(record.text, 120);
        if (text && text !== '(empty)') textParts.push(text);
        continue;
      }
      if (type === 'tool_use' && typeof record.name === 'string') {
        const name = record.name.trim();
        if (name && !toolNames.includes(name)) toolNames.push(name);
        continue;
      }
      if (type === 'tool_result' && textParts.length === 0 && typeof record.content === 'string') {
        const text = summarizePlainContextText(record.content, 80);
        if (text && text !== '(empty)') textParts.push(text);
      }
    }

    const parts: string[] = [];
    if (textParts.length > 0) parts.push(textParts[0]);
    if (toolNames.length > 0) {
      const toolSummary = toolNames.length <= 2
        ? toolNames.join(' + ')
        : `${toolNames.slice(0, 2).join(' + ')} +${toolNames.length - 2}`;
      parts.push(toolSummary);
    }

    if (parts.length === 0) return null;
    return summarizePlainContextText(parts.join(' · '), maxLength);
  } catch {
    return null;
  }
}

function summarizeContextText(content: string, maxLength = 160): string {
  return summarizeStructuredContextText(content, maxLength) ?? summarizePlainContextText(content, maxLength);
}

function renderStructuredContextDetail(content: string, maxLength = 2400): string | null {
  const trimmed = content.trim();
  if (!trimmed || (!trimmed.startsWith('[') && !trimmed.startsWith('{'))) return null;

  try {
    const parsed = JSON.parse(trimmed);
    const blocks = Array.isArray(parsed) ? parsed : [parsed];
    if (!blocks.some((block) => block && typeof block === 'object' && typeof block.type === 'string')) {
      return null;
    }

    const parts: string[] = [];
    for (const block of blocks) {
      if (!block || typeof block !== 'object') continue;
      const record = block as Record<string, unknown>;
      const type = typeof record.type === 'string' ? record.type : '';
      if (type === 'text' && typeof record.text === 'string' && record.text.trim()) {
        parts.push(record.text.trim());
        continue;
      }
      if (type === 'tool_use' && typeof record.name === 'string') {
        parts.push(`**Tool** \`${escapeCardText(record.name.trim())}\``);
        continue;
      }
      if (type === 'tool_result' && typeof record.content === 'string' && record.content.trim()) {
        parts.push(`**Result** ${escapeCardText(summarizePlainContextText(record.content, 240))}`);
      }
    }

    if (parts.length === 0) return null;
    const merged = parts.join('\n\n').trim();
    return merged.length > maxLength ? `${merged.slice(0, maxLength - 3)}...` : merged;
  } catch {
    return null;
  }
}

function renderContextDetail(content: string, maxLength = 2400): string {
  return renderStructuredContextDetail(content, maxLength) ?? summarizePlainContextText(content, maxLength);
}

function buildContextMessagePanel(message: BridgeMessage, index: number): Record<string, unknown> {
  return {
    tag: 'collapsible_panel',
    expanded: false,
    border: {
      color: 'grey',
      corner_radius: '6px',
    },
    header: {
      title: {
        tag: 'markdown',
        content: `**${index + 1}. ${getRoleLabel(message.role)}** · ${escapeCardText(summarizeContextText(message.content, 84))}`,
      },
      icon_position: 'follow_text',
      icon_expanded_angle: -180,
    },
    padding: '6px',
    vertical_spacing: '4px',
    elements: [{
      tag: 'markdown',
      content: renderContextDetail(message.content, 1200),
      text_size: 'normal',
    }],
  };
}

function chunkCardButtons<T>(items: T[], size = 5): T[][] {
  const chunks: T[][] = [];
  for (let index = 0; index < items.length; index += size) {
    chunks.push(items.slice(index, index + size));
  }
  return chunks;
}

function getRoleLabel(role: string): string {
  const normalized = role.trim().toLowerCase();
  if (normalized === 'assistant') return 'Assistant';
  if (normalized === 'user') return 'User';
  if (normalized === 'system') return 'System';
  if (normalized === 'tool') return 'Tool';
  return role || 'Message';
}

function resolveDisplaySession(sessionId: string): NamedBridgeSession | null {
  const store = getSessionListingStore();
  const direct = store.getSession(sessionId) as NamedBridgeSession | null;
  if (direct) return direct;

  const discovered = discoverCodexSessions().find((session) => session.id === sessionId);
  if (discovered) return discovered;

  return listBridgeSessions().find((session) => getStoredSdkSessionId(session) === sessionId) ?? null;
}

function getNativeSessionLogPath(session: NamedBridgeSession | null | undefined): string | null {
  if (!session) return null;
  if (typeof session.native_log_path === 'string' && session.native_log_path) {
    return session.native_log_path;
  }

  const nativeSessionId = getStoredSdkSessionId(session);
  if (!nativeSessionId) return null;
  const discovered = discoverCodexSessions().find((item) => item.id === nativeSessionId);
  return typeof discovered?.native_log_path === 'string' ? discovered.native_log_path : null;
}

function getNativeSessionMessages(session: NamedBridgeSession | null | undefined): BridgeMessage[] {
  const logPath = getNativeSessionLogPath(session);
  if (!logPath) return [];

  const mtimeMs = safeStatMtimeMs(logPath);
  const cached = nativeSessionMessageCache.get(logPath);
  if (cached && cached.mtimeMs === mtimeMs) {
    return cached.messages;
  }

  const messages: BridgeMessage[] = [];
  try {
    const fileText = fs.readFileSync(logPath, 'utf8');
    for (const line of fileText.split(/\r?\n/)) {
      if (!line.trim()) continue;
      try {
        const item = JSON.parse(line) as {
          type?: unknown;
          payload?: { type?: unknown; message?: unknown };
        };
        if (item.type !== 'event_msg') continue;
        const payloadType = item.payload?.type;
        const message = item.payload?.message;
        if (payloadType === 'user_message' && typeof message === 'string' && message.trim()) {
          messages.push({ role: 'user', content: message });
        } else if (payloadType === 'agent_message' && typeof message === 'string' && message.trim()) {
          messages.push({ role: 'assistant', content: message });
        }
      } catch {
        // Ignore partially written rows while Codex is still appending to the log.
      }
    }
  } catch {
    return [];
  }

  nativeSessionMessageCache.set(logPath, { mtimeMs, messages });
  return messages;
}

function getEffectiveSessionMessages(sessionId: string): { session: NamedBridgeSession; messages: BridgeMessage[] } | null {
  const session = resolveDisplaySession(sessionId);
  if (!session) return null;

  const store = getSessionListingStore();
  const localSession = store.getSession(sessionId) as NamedBridgeSession | null;
  const localMessages = localSession ? store.getMessages(sessionId).messages : [];
  if (localMessages.length > 0) {
    return { session, messages: localMessages };
  }

  const nativeMessages = getNativeSessionMessages(session);
  return { session, messages: nativeMessages };
}

function getSessionContextPreview(sessionId: string, page = 0, pageSize = CONTEXT_PAGE_SIZE): SessionContextPreview | null {
  const effective = getEffectiveSessionMessages(sessionId);
  if (!effective) return null;
  const { session, messages: allMessages } = effective;
  const totalMessages = allMessages.length;
  const safePageSize = Math.max(1, pageSize);
  const totalPages = Math.max(1, Math.ceil(totalMessages / safePageSize));
  const safePage = Math.min(Math.max(0, page), totalPages - 1);
  const sliceEnd = totalMessages - safePage * safePageSize;
  const sliceStart = Math.max(0, sliceEnd - safePageSize);
  const messages = allMessages.slice(sliceStart, Math.max(sliceStart, sliceEnd));
  return {
    session,
    messages,
    page: safePage,
    pageSize: safePageSize,
    totalMessages,
    totalPages,
    startIndex: totalMessages === 0 ? 0 : sliceStart + 1,
    endIndex: totalMessages === 0 ? 0 : sliceStart + messages.length,
  };
}

function getSessionMessageCount(sessionId: string): number {
  return getEffectiveSessionMessages(sessionId)?.messages.length ?? 0;
}

function buildContextPanel(
  sessionId: string,
  page = 0,
  pagerMode: 'status' | 'peek' = 'peek',
  detailIndex?: number,
): Record<string, unknown> {
  const preview = getSessionContextPreview(sessionId, page);
  const pagerButtons: Array<{ label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> }> = [];
  const selectedDetail = preview && detailIndex != null && detailIndex >= 0 && detailIndex < preview.messages.length
    ? preview.messages[detailIndex]
    : null;

  const buildPager = (targetPage: number): { callbackData: string; extras: Record<string, string> } => {
    if (pagerMode === 'status') {
      return {
        callbackData: `nav:status:${targetPage}`,
        extras: {
          nav: 'status',
          page: String(targetPage),
        },
      };
    }
    return {
      callbackData: `nav:peek:${sessionId}:${targetPage}`,
      extras: {
        nav: 'peek',
        session_id: sessionId,
        page: String(targetPage),
      },
    };
  };

  const buildCloseDetail = (): { callbackData: string; extras: Record<string, string> } => {
    if (pagerMode === 'status') {
      return {
        callbackData: `nav:ctxclear:status:${page}`,
        extras: {
          nav: 'ctxclear_status',
          page: String(page),
        },
      };
    }
    return {
      callbackData: `nav:ctxclear:peek:${sessionId}:${page}`,
      extras: {
        nav: 'ctxclear_peek',
        session_id: sessionId,
        page: String(page),
      },
    };
  };

  if (preview && preview.page > 0) {
    const pager = buildPager(preview.page - 1);
    pagerButtons.push({
      label: 'Newer',
      callbackData: pager.callbackData,
      type: 'default',
      extras: pager.extras,
    });
  }
  if (preview && preview.page < preview.totalPages - 1) {
    const pager = buildPager(preview.page + 1);
    pagerButtons.push({
      label: 'Older',
      callbackData: pager.callbackData,
      type: 'default',
      extras: pager.extras,
    });
  }

  const elements: Array<Record<string, unknown>> = [];
  const latestMessage = preview?.messages.at(-1) ?? null;
  const latestSummary = latestMessage
    ? `${getRoleLabel(latestMessage.role)}: ${summarizeContextText(latestMessage.content, 40)}`
    : '';
  if (selectedDetail) {
    const closeDetail = buildCloseDetail();
    elements.push({
      tag: 'markdown',
      content: `**${getRoleLabel(selectedDetail.role)}** · full content`,
      text_size: 'normal',
    });
    elements.push({
      tag: 'markdown',
      content: renderContextDetail(selectedDetail.content),
      text_size: 'normal',
    });
    elements.push(createInlineCardButtonSet([{
      label: 'Close',
      callbackData: closeDetail.callbackData,
      type: 'default',
      extras: closeDetail.extras,
    }]));
  } else if (preview?.messages.length) {
    preview.messages.forEach((message, index) => {
      elements.push(buildContextMessagePanel(message, index));
    });
  } else {
    elements.push({
      tag: 'markdown',
      content: 'No saved messages yet.',
      text_size: 'normal',
    });
  }

  return {
    tag: 'collapsible_panel',
    expanded: false,
    border: {
      color: 'grey',
      corner_radius: '6px',
    },
    header: {
      title: {
        tag: 'markdown',
        content: `**Recent Context**${preview ? ` · ${preview.startIndex}-${preview.endIndex} / ${preview.totalMessages} msg` : ''}${latestSummary ? ` · ${escapeCardText(latestSummary)}` : ''}`,
      },
      icon_position: 'follow_text',
      icon_expanded_angle: -180,
    },
    padding: '8px',
    vertical_spacing: '6px',
    elements: [
      ...elements,
      ...(pagerButtons.length > 0 ? [createInlineCardButtonSet(pagerButtons)] : []),
    ],
  };
}

function encodeCardToken(value: string): string {
  return Buffer.from(value, 'utf8').toString('base64url');
}

function decodeCardToken(value: string): string | null {
  try {
    return Buffer.from(value, 'base64url').toString('utf8');
  } catch {
    return null;
  }
}

function resolveWorkspaceFilter(
  currentBinding: ChannelBinding,
  scope: ProjectListScope = 'global',
  workspacePath?: string | null,
): string | null {
  if (workspacePath?.trim()) {
    const normalized = normalizeProjectPath(workspacePath);
    return isKnownWorkspaceRoot(normalized) ? normalized : null;
  }
  if (scope === 'workspace') {
    return getContainingWorkspaceRoot(currentBinding.workingDirectory);
  }
  return null;
}

function buildWorkspaceGroups(currentBinding: ChannelBinding): WorkspaceGroup[] {
  currentBinding = syncBindingToCurrentSession(currentBinding);
  const workspaces = new Map<string, { projects: Set<string>; sessionCount: number }>();
  const currentWorkspaceRoot = getContainingWorkspaceRoot(currentBinding.workingDirectory);

  for (const session of listKnownSessions()) {
    const projectPath = normalizeProjectPath(session.working_directory);
    const workspaceRoot = getContainingWorkspaceRoot(projectPath);
    if (!workspaceRoot) {
      continue;
    }
    const group = workspaces.get(workspaceRoot) ?? { projects: new Set<string>(), sessionCount: 0 };
    group.projects.add(projectPath);
    group.sessionCount += 1;
    workspaces.set(workspaceRoot, group);
  }

  return Array.from(workspaces.entries())
    .map(([workspacePath, stats]) => ({
      path: workspacePath,
      label: getWorkspaceLabel(workspacePath),
      projectCount: stats.projects.size,
      sessionCount: stats.sessionCount,
      current: workspacePath === currentWorkspaceRoot,
    }))
    .sort((left, right) => {
      if (left.current !== right.current) return left.current ? -1 : 1;
      return left.label.localeCompare(right.label, 'en');
    });
}

function buildProjectGroups(
  currentBinding: ChannelBinding,
  scope: ProjectListScope = 'global',
  workspacePath?: string | null,
): ProjectGroup[] {
  currentBinding = syncBindingToCurrentSession(currentBinding);
  const groups = new Map<string, ProjectGroup>();
  const currentProjectPath = normalizeProjectPath(currentBinding.workingDirectory);
  const workspaceRoot = resolveWorkspaceFilter(currentBinding, scope, workspacePath);

  for (const session of listKnownSessions()) {
    const projectPath = normalizeProjectPath(session.working_directory);
    if (workspaceRoot && !isInWorkspace(projectPath, workspaceRoot)) {
      continue;
    }
    let group = groups.get(projectPath);
    if (!group) {
      group = {
        path: projectPath,
        label: getProjectLabel(projectPath),
        sessions: [],
        current: projectPath === currentProjectPath,
      };
      groups.set(projectPath, group);
    }
    group.sessions.push(session);
  }

  return Array.from(groups.values())
    .sort((left, right) => {
      if (left.current !== right.current) return left.current ? -1 : 1;
      return left.label.localeCompare(right.label, 'en');
    })
    .map((group) => ({
      ...group,
      sessions: [...group.sessions].sort((left, right) => {
        const sortTime = getSessionSortTime(right) - getSessionSortTime(left);
        return sortTime || right.id.localeCompare(left.id, 'en');
      }),
    }));
}

function escapeCardText(value: string): string {
  return value.replace(/[`]/g, '\\`');
}

function createButtonValue(
  callbackData: string,
  extras?: Record<string, string>,
): Record<string, string> {
  return {
    callback_data: callbackData,
    ...(extras ?? {}),
  };
}

function createCardButtonRow(
  label: string,
  callbackData: string,
  type: 'default' | 'primary' = 'default',
  extras?: Record<string, string>,
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: label.slice(0, 60) },
            type,
            size: 'medium',
            value: createButtonValue(callbackData, extras),
          },
        ],
      },
    ],
  };
}

function createInlineCardButtonSet(
  buttons: Array<{ label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> }>,
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: buttons.map((button, index) => ({
      tag: 'column',
      width: 'auto',
      elements: [
        {
          tag: 'button',
          text: { tag: 'plain_text', content: button.label.slice(0, 24) },
          type: button.type ?? 'default',
          size: 'small',
          width: 'default',
          margin: index < buttons.length - 1 ? '0px 8px 0px 0px' : '0px 0px 0px 0px',
          value: createButtonValue(button.callbackData, button.extras),
        },
      ],
    })),
  };
}

function createCardButtonGrid(
  buttons: Array<{ label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> }>,
  columnsPerRow = 2,
): Array<Record<string, unknown>> {
  const rows: Array<Record<string, unknown>> = [];
  for (const row of chunkCardButtons(buttons, columnsPerRow)) {
    rows.push({
      tag: 'column_set',
      flex_mode: 'none',
      columns: row.map((button, index) => ({
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: button.label.slice(0, 24) },
            type: button.type ?? 'default',
            size: 'small',
            width: 'fill',
            margin: index < row.length - 1 ? '0px 6px 0px 0px' : '0px 0px 0px 0px',
            value: createButtonValue(button.callbackData, button.extras),
          },
        ],
      })),
    });
  }
  return rows;
}

function getWorkspaceButtonLabel(workspacePath: string): string {
  const label = getWorkspaceLabel(workspacePath).replace(/\s+/g, ' ').trim() || getCompactProjectPath(workspacePath);
  return label.length > 14 ? `${label.slice(0, 12)}...` : label;
}

function createWorkspaceCallback(workspacePath: string | null): string {
  return workspacePath ? `nav:workspace:${encodeCardToken(workspacePath)}` : 'nav:projects:global';
}

function createWorkspaceExtras(workspacePath: string | null): Record<string, string> {
  if (!workspacePath) {
    return { nav: 'projects', scope: 'global' };
  }
  return {
    nav: 'workspace',
    workspace_path: workspacePath,
  };
}

function createProjectNavCallback(
  projectPath: string,
  workspacePath?: string | null,
  page = 0,
  scope: ProjectListScope = 'workspace',
): string {
  const projectToken = encodeCardToken(projectPath);
  const pageSuffix = page > 0 ? `:${page}` : '';
  return workspacePath
    ? `nav:project:${projectToken}:${encodeCardToken(workspacePath)}${pageSuffix}`
    : `nav:project:${scope}:${projectToken}${pageSuffix}`;
}

function createProjectNavExtras(
  projectPath: string,
  workspacePath?: string | null,
  page = 0,
  scope: ProjectListScope = 'workspace',
): Record<string, string> {
  return {
    nav: 'project',
    project_path: projectPath,
    ...(!workspacePath ? { scope } : {}),
    ...(page > 0 ? { page: String(page) } : {}),
    ...(workspacePath ? { workspace_path: workspacePath } : {}),
  };
}

function createProjectNewCallback(
  projectPath: string,
  workspacePath?: string | null,
  scope: ProjectListScope = 'workspace',
): string {
  const projectToken = encodeCardToken(projectPath);
  return workspacePath
    ? `nav:new:${projectToken}:${encodeCardToken(workspacePath)}`
    : `nav:new:${scope}:${projectToken}`;
}

function createProjectNewExtras(
  projectPath: string,
  workspacePath?: string | null,
  scope: ProjectListScope = 'workspace',
): Record<string, string> {
  return {
    nav: 'project_new',
    project_path: projectPath,
    ...(!workspacePath ? { scope } : {}),
    ...(workspacePath ? { workspace_path: workspacePath } : {}),
  };
}

function createDockButtonRow(
  primary: { label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> },
  close?: { callbackData: string; extras?: Record<string, string> },
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: primary.label.slice(0, 48) },
            type: primary.type ?? 'default',
            size: 'small',
            width: 'fill',
            value: createButtonValue(primary.callbackData, primary.extras),
          },
        ],
      },
      ...(close ? [{
        tag: 'column',
        width: 'auto',
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: '×' },
            type: 'default',
            size: 'small',
            width: 'default',
            margin: '0px 0px 0px 6px',
            value: createButtonValue(close.callbackData, close.extras),
          },
        ],
      }] : []),
    ],
  };
}

function createProjectSummaryNavRow(
  projectLabel: string,
  sessionId: string,
  compactProjectPath: string,
  projectPath: string,
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [
          {
            tag: 'markdown',
            content: [
              `**${escapeCardText(projectLabel)}** · \`${getShortSessionId(sessionId)}\``,
              `\`${escapeCardText(compactProjectPath)}\``,
            ].join('\n'),
          },
        ],
      },
      {
        tag: 'column',
        width: 'auto',
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: 'Sessions' },
            type: 'primary',
            size: 'small',
            width: 'default',
            value: createButtonValue(
              createProjectNavCallback(projectPath, getContainingWorkspaceRoot(projectPath), 0, getProjectNavigationScope(projectPath)),
              createProjectNavExtras(projectPath, getContainingWorkspaceRoot(projectPath), 0, getProjectNavigationScope(projectPath)),
            ),
          },
        ],
      },
      {
        tag: 'column',
        width: 'auto',
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: 'Projects' },
            type: 'default',
            size: 'small',
            width: 'default',
            margin: '0px 0px 0px 6px',
            value: createButtonValue('nav:projects', {
              nav: 'projects',
            }),
          },
        ],
      },
    ],
  };
}

function createRightAlignedButtonRow(
  label: string,
  callbackData: string,
  type: 'default' | 'primary' = 'default',
  extras?: Record<string, string>,
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [],
      },
      {
        tag: 'column',
        width: 'auto',
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: label.slice(0, 20) },
            type,
            size: 'small',
            width: 'default',
            value: createButtonValue(callbackData, extras),
          },
        ],
      },
    ],
  };
}

function createSummaryActionRow(
  summary: string,
  label: string,
  callbackData: string,
  type: 'default' | 'primary' = 'default',
  extras?: Record<string, string>,
): Record<string, unknown> {
  return {
    tag: 'column_set',
    flex_mode: 'none',
    columns: [
      {
        tag: 'column',
        width: 'weighted',
        weight: 1,
        elements: [
          {
            tag: 'markdown',
            content: summary,
            text_size: 'notation',
          },
        ],
      },
      {
        tag: 'column',
        width: 'auto',
        elements: [
          {
            tag: 'button',
            text: { tag: 'plain_text', content: label.slice(0, 20) },
            type,
            size: 'small',
            width: 'default',
            value: createButtonValue(callbackData, extras),
          },
        ],
      },
    ],
  };
}

function buildSessionDockPanel(
  binding: ChannelBinding,
  expanded = true,
  options?: {
    bordered?: boolean;
    padding?: string;
    verticalSpacing?: string;
    title?: string;
  },
): Record<string, unknown> {
  const items = getSessionDockItems(binding);
  const runningCount = items.filter((item) => item.running).length;
  const unreadCount = items.reduce((total, item) => total + (item.current ? 0 : item.unreadCount), 0);
  const projectCount = new Set(items.map((item) => item.projectLabel)).size;
  const elements: Array<Record<string, unknown>> = [];
  const bordered = options?.bordered ?? true;
  const padding = options?.padding ?? '8px';
  const verticalSpacing = options?.verticalSpacing ?? '4px';
  const dockSummary = `${items.length} open${projectCount > 1 ? ` · ${projectCount} projects` : ''}${runningCount ? ` · ${runningCount} running` : ''}${unreadCount ? ` · ${unreadCount} unread` : ''}`;
  const title = options?.title
    ?? `**Open Sessions** · ${dockSummary}`;

  if (items.length === 0) {
    elements.push({
      tag: 'markdown',
      content: 'No opened sessions yet.',
      text_size: 'normal',
    });
  } else {
    items.forEach((item) => {
      const indicator = getDockIndicator(item);
      const statusParts = [
        item.current ? 'current' : null,
        item.running ? 'run' : null,
        item.unreadCount > 0 && !item.current ? `+${item.unreadCount}` : null,
      ].filter(Boolean);
      const statusSuffix = statusParts.length > 0 ? ` · ${statusParts.join(' · ')}` : '';
      const title = `${indicator} ${item.projectLabel}/${getShortSessionId(item.sessionId)} · ${item.messageCount} msg${statusSuffix}`;
      elements.push(createDockButtonRow(
        {
          label: title,
          callbackData: `nav:dock:select:${item.sessionId}`,
          type: item.current ? 'primary' : 'default',
          extras: {
            nav: 'dock_select',
            session_id: item.sessionId,
          },
        },
        items.length > 1 ? {
          callbackData: `nav:dock:close:${item.sessionId}`,
          extras: {
            nav: 'dock_close',
            session_id: item.sessionId,
          },
        } : undefined,
      ));
    });
  }

  return {
    tag: 'collapsible_panel',
    expanded,
    ...(bordered ? {
      border: {
        color: 'grey',
        corner_radius: '6px',
      },
    } : {}),
    header: {
      title: {
        tag: 'markdown',
        content: title,
      },
      icon_position: 'follow_text',
      icon_expanded_angle: -180,
    },
    padding,
    vertical_spacing: verticalSpacing,
    elements,
  };
}

function buildFeishuProjectListCard(
  groups: ProjectGroup[],
  binding: ChannelBinding,
  scope: ProjectListScope = 'global',
  workspacePath?: string | null,
): string {
  binding = syncBindingToCurrentSession(binding);
  const selectedWorkspace = resolveWorkspaceFilter(binding, scope, workspacePath);
  const workspaceLabel = selectedWorkspace ? getWorkspaceLabel(selectedWorkspace) : 'All Workspaces';
  const workspaceGroups = buildWorkspaceGroups(binding);
  const workspaceButtons = [
    {
      label: selectedWorkspace ? 'All' : 'All *',
      callbackData: createWorkspaceCallback(null),
      type: selectedWorkspace ? 'default' as const : 'primary' as const,
      extras: createWorkspaceExtras(null),
    },
    ...workspaceGroups.slice(0, 7).map((workspace) => ({
      label: `${getWorkspaceButtonLabel(workspace.path)}${selectedWorkspace === workspace.path ? ' *' : ''}`,
      callbackData: createWorkspaceCallback(workspace.path),
      type: selectedWorkspace === workspace.path ? 'primary' as const : 'default' as const,
      extras: createWorkspaceExtras(workspace.path),
    })),
  ];
  const elements: Array<Record<string, unknown>> = [
    buildSessionDockPanel(binding, false, {
      bordered: false,
      padding: '0px',
      verticalSpacing: '2px',
      title: `**Open Sessions** · ${getOpenSessionIds(binding).length} open`,
    }),
    { tag: 'hr' },
    buildCurrentSessionPanel(binding, {
      buttons: [
        {
          label: 'Ctx',
          callbackData: `nav:peek:${binding.codepilotSessionId}`,
          type: 'default',
          extras: {
            nav: 'peek',
            session_id: binding.codepilotSessionId,
          },
        },
        {
          label: 'Current Project',
          callbackData: createProjectNavCallback(
            normalizeProjectPath(binding.workingDirectory),
            getContainingWorkspaceRoot(binding.workingDirectory),
            0,
            getProjectNavigationScope(binding.workingDirectory),
          ),
          type: 'primary',
          extras: createProjectNavExtras(
            normalizeProjectPath(binding.workingDirectory),
            getContainingWorkspaceRoot(binding.workingDirectory),
            0,
            getProjectNavigationScope(binding.workingDirectory),
          ),
        },
      ],
    }),
    { tag: 'hr' },
    {
      tag: 'markdown',
      content: `**Workspace** · ${escapeCardText(workspaceLabel)} · ${formatProjectCount(groups.length)}`,
      text_size: 'notation',
    },
    ...createCardButtonGrid(workspaceButtons, 2),
  ];

  if (groups.length === 0) {
    elements.push({ tag: 'hr' });
    elements.push({
      tag: 'markdown',
      content: 'No projects found yet.',
    });
  } else {
    groups.slice(0, 12).forEach((group, index) => {
      if (index === 0) {
        elements.push({ tag: 'hr' });
      }
      elements.push(createSummaryActionRow(
        [
          `**${escapeCardText(group.label)}**${group.current ? ' · Current' : ''}`,
          `\`${escapeCardText(getProjectPathHint(group.path))}\` · ${formatSessionCount(group.sessions.length)}`,
        ].join('\n'),
        'Open',
        createProjectNavCallback(group.path, selectedWorkspace, 0, scope),
        group.current ? 'primary' : 'default',
        createProjectNavExtras(group.path, selectedWorkspace, 0, scope),
      ));
      if (index < groups.length - 1) {
        elements.push({ tag: 'hr' });
      }
    });
  }

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    header: {
      title: { tag: 'plain_text', content: 'Projects' },
      template: 'default',
      text_tag_list: [
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: selectedWorkspace ? 'Workspace' : 'All',
          },
          color: selectedWorkspace ? 'blue' : 'wathet',
        },
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: formatProjectCount(groups.length),
          },
          color: 'grey',
        },
      ],
    },
    body: { elements },
  });
}

function buildFeishuProjectSessionsCard(
  group: ProjectGroup,
  currentSessionId: string,
  binding?: ChannelBinding,
  scope: ProjectListScope = 'workspace',
  workspacePath?: string | null,
  page = 0,
): string {
  binding = binding ? syncBindingToCurrentSession(binding) : binding;
  const selectedWorkspace = binding
    ? resolveWorkspaceFilter(binding, scope, workspacePath)
    : (workspacePath ? normalizeProjectPath(workspacePath) : null);
  const totalPages = Math.max(1, Math.ceil(group.sessions.length / PROJECT_SESSIONS_PAGE_SIZE));
  const safePage = Math.min(Math.max(0, page), totalPages - 1);
  const startIndex = safePage * PROJECT_SESSIONS_PAGE_SIZE;
  const sessionItems: FeishuSessionCardItem[] = group.sessions
    .slice(startIndex, startIndex + PROJECT_SESSIONS_PAGE_SIZE)
    .map((session) => ({
    sessionId: session.id,
    title: getSessionTitle(session),
    subtitle: `${getShortSessionId(session.id)} · ${formatMessageCount(getSessionMessageCount(session.id))} · ${session.model || 'default'}`,
    timestamps: buildSessionTimestampSummary(session),
    current: session.id === currentSessionId,
    messageCount: getSessionMessageCount(session.id),
  }));

  const elements: Array<Record<string, unknown>> = [
    createSummaryActionRow(
      `Back to ${escapeCardText(selectedWorkspace ? getWorkspaceLabel(selectedWorkspace) : 'all workspaces')}`,
      '◂ Projects',
      selectedWorkspace ? createWorkspaceCallback(selectedWorkspace) : `nav:projects:${scope}`,
      'default',
      selectedWorkspace ? createWorkspaceExtras(selectedWorkspace) : {
        nav: 'projects',
        scope,
      },
    ),
    ...(binding ? [buildSessionDockPanel(binding, false, {
      bordered: false,
      padding: '0px',
      verticalSpacing: '2px',
      title: `**Open Sessions** · ${getOpenSessionIds(binding).length} open`,
    })] : []),
    { tag: 'hr' },
    createInlineCardButtonSet([
      {
        label: 'New',
        callbackData: createProjectNewCallback(group.path, selectedWorkspace, scope),
        type: 'primary',
        extras: createProjectNewExtras(group.path, selectedWorkspace, scope),
      },
      {
        label: 'Status',
        callbackData: 'nav:status',
        type: 'default',
        extras: {
          nav: 'status',
        },
      },
    ]),
    { tag: 'hr' },
  ];

  if (sessionItems.length === 0) {
    elements.push({
      tag: 'markdown',
      content: 'No sessions found for this project.',
    });
  } else {
    sessionItems.forEach((session, index) => {
      const lines = [
        `**${escapeCardText(session.title)}**`,
        `\`${escapeCardText(session.subtitle)}\``,
        escapeCardText(session.timestamps),
      ];
      elements.push({
        tag: 'markdown',
        content: lines.join('\n'),
      });
      elements.push(createInlineCardButtonSet([
        {
          label: session.current ? 'Current' : 'Use',
          callbackData: `nav:bind:${session.sessionId}`,
          type: session.current ? 'primary' : 'default',
          extras: {
            nav: 'bind',
            session_id: session.sessionId,
          },
        },
        {
          label: 'Ctx',
          callbackData: `nav:peek:${session.sessionId}`,
          type: 'default',
          extras: {
            nav: 'peek',
            session_id: session.sessionId,
          },
        },
        {
          label: 'Arc',
          callbackData: `nav:archive:${session.sessionId}`,
          type: 'default',
          extras: {
            nav: 'archive',
            session_id: session.sessionId,
          },
        },
      ]));
      if (index < sessionItems.length - 1) {
        elements.push({ tag: 'hr' });
      }
    });
  }

  const pagerButtons: Array<{ label: string; callbackData: string; type?: 'default' | 'primary'; extras?: Record<string, string> }> = [];
  if (safePage > 0) {
    pagerButtons.push({
      label: 'Newer',
      callbackData: createProjectNavCallback(group.path, selectedWorkspace, safePage - 1, scope),
      type: 'default',
      extras: createProjectNavExtras(group.path, selectedWorkspace, safePage - 1, scope),
    });
  }
  if (safePage < totalPages - 1) {
    pagerButtons.push({
      label: 'Older',
      callbackData: createProjectNavCallback(group.path, selectedWorkspace, safePage + 1, scope),
      type: 'default',
      extras: createProjectNavExtras(group.path, selectedWorkspace, safePage + 1, scope),
    });
  }
  if (pagerButtons.length > 0) {
    elements.push({ tag: 'hr' });
    elements.push(createInlineCardButtonSet(pagerButtons));
  }

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    header: {
      title: { tag: 'plain_text', content: group.label },
      template: 'default',
      text_tag_list: [
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: selectedWorkspace ? getWorkspaceLabel(selectedWorkspace) : (scope === 'workspace' ? 'Workspace' : 'All'),
          },
          color: selectedWorkspace || scope === 'workspace' ? 'blue' : 'wathet',
        },
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: formatSessionCount(group.sessions.length),
          },
          color: 'blue',
        },
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: `${safePage + 1}/${totalPages}`,
          },
          color: 'grey',
        },
      ],
    },
    body: { elements },
  });
}

function buildFeishuStatusCard(binding: ChannelBinding, contextPage = 0, contextDetailIndex?: number): string {
  binding = syncBindingToCurrentSession(binding);
  const projectPath = normalizeProjectPath(binding.workingDirectory);
  const projectLabel = getProjectLabel(projectPath);
  const compactProjectPath = getCompactProjectPath(projectPath);
  const messageCount = getSessionMessageCount(binding.codepilotSessionId);

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    header: {
      title: { tag: 'plain_text', content: 'Current Session' },
      template: 'wathet',
      text_tag_list: [
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: formatMessageCount(messageCount),
          },
          color: 'blue',
        },
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: binding.mode,
          },
          color: 'wathet',
        },
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: binding.model || 'default',
          },
          color: 'grey',
        },
      ],
    },
    body: {
      elements: [
        createProjectSummaryNavRow(projectLabel, binding.codepilotSessionId, compactProjectPath, projectPath),
        buildSessionDockPanel(binding),
        buildContextPanel(binding.codepilotSessionId, contextPage, 'status', contextDetailIndex),
      ],
    },
  });
}

function buildFeishuSessionPreviewCard(sessionId: string, currentSessionId: string, contextPage = 0, contextDetailIndex?: number): string | null {
  const preview = getSessionContextPreview(sessionId, contextPage);
  if (!preview) return null;

  const projectPath = normalizeProjectPath(preview.session.working_directory);
  const projectLabel = getProjectLabel(projectPath);
  const compactProjectPath = getCompactProjectPath(projectPath);
  const title = getSessionTitle(preview.session);
  const messageCount = getSessionMessageCount(sessionId);

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    header: {
      title: { tag: 'plain_text', content: title },
      template: 'default',
      text_tag_list: [
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: `${formatMessageCount(messageCount)} / ${preview.session.model || 'default'}`,
          },
          color: 'blue',
        },
        ...(sessionId === currentSessionId
          ? [{
              tag: 'text_tag',
              text: {
                tag: 'plain_text',
                content: 'Current',
              },
              color: 'green',
            }]
          : []),
      ],
    },
    body: {
      elements: [
        {
          tag: 'markdown',
          content: `**${escapeCardText(projectLabel)}**\n\`${escapeCardText(compactProjectPath)}\``,
        },
        createInlineCardButtonSet([
          {
            label: sessionId === currentSessionId ? 'Current' : 'Use',
            callbackData: `nav:bind:${sessionId}`,
            type: sessionId === currentSessionId ? 'primary' : 'default',
            extras: {
              nav: 'bind',
              session_id: sessionId,
            },
          },
          {
            label: 'Back',
            callbackData: createProjectNavCallback(
              projectPath,
              getContainingWorkspaceRoot(projectPath),
              0,
              getProjectNavigationScope(projectPath),
            ),
            type: 'default',
            extras: createProjectNavExtras(
              projectPath,
              getContainingWorkspaceRoot(projectPath),
              0,
              getProjectNavigationScope(projectPath),
            ),
          },
          {
            label: 'Archive',
            callbackData: `nav:archive:${sessionId}`,
            type: 'default',
            extras: {
              nav: 'archive',
              session_id: sessionId,
            },
          },
        ]),
        buildContextPanel(sessionId, contextPage, 'peek', contextDetailIndex),
      ],
    },
  });
}

function buildFeishuReplyNavCard(binding: ChannelBinding): string {
  binding = syncBindingToCurrentSession(binding);
  const projectPath = normalizeProjectPath(binding.workingDirectory);

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    body: {
      elements: [
        buildCurrentSessionPanel(binding, {
          showTimestamps: false,
          buttons: [
            {
              label: 'Ctx',
              callbackData: `nav:peek:${binding.codepilotSessionId}`,
              type: 'default',
              extras: {
                nav: 'peek',
                session_id: binding.codepilotSessionId,
              },
            },
          ],
        }),
        {
          tag: 'column_set',
          flex_mode: 'none',
          columns: [
            {
              tag: 'column',
              width: 'weighted',
              weight: 1,
              elements: [
                {
                  tag: 'button',
                  text: { tag: 'plain_text', content: 'Current Project' },
                  type: 'primary',
                  size: 'small',
                  value: createButtonValue(
                    createProjectNavCallback(projectPath, getContainingWorkspaceRoot(projectPath), 0, getProjectNavigationScope(projectPath)),
                    createProjectNavExtras(projectPath, getContainingWorkspaceRoot(projectPath), 0, getProjectNavigationScope(projectPath)),
                  ),
                },
              ],
            },
            {
              tag: 'column',
              width: 'weighted',
              weight: 1,
              elements: [
                {
                  tag: 'button',
                  text: { tag: 'plain_text', content: 'All Projects' },
                  type: 'default',
                  size: 'small',
                  value: createButtonValue('nav:projects', {
                    nav: 'projects',
                  }),
                },
              ],
            },
          ],
        },
      ],
    },
  });
}

function buildFeishuResponseCard(binding: ChannelBinding, responseText: string): string {
  binding = syncBindingToCurrentSession(binding);
  const projectPath = normalizeProjectPath(binding.workingDirectory);
  const projectLabel = getProjectLabel(projectPath);
  const sessionLabel = getCurrentSessionLabel(binding);

  return JSON.stringify({
    schema: '2.0',
    config: { wide_screen_mode: true, update_multi: true },
    header: {
      title: {
        tag: 'plain_text',
        content: projectLabel,
      },
      template: 'default',
      text_tag_list: [
        {
          tag: 'text_tag',
          text: {
            tag: 'plain_text',
            content: sessionLabel,
          },
          color: 'blue',
        },
      ],
    },
    body: {
      elements: [
        buildCurrentSessionPanel(binding, {
          showTimestamps: false,
          buttons: [
            {
              label: 'Ctx',
              callbackData: `nav:peek:${binding.codepilotSessionId}`,
              type: 'default',
              extras: {
                nav: 'peek',
                session_id: binding.codepilotSessionId,
              },
            },
            {
              label: 'Sessions',
              callbackData: createProjectNavCallback(
                projectPath,
                getContainingWorkspaceRoot(projectPath),
                0,
                getProjectNavigationScope(projectPath),
              ),
              type: 'primary',
              extras: createProjectNavExtras(
                projectPath,
                getContainingWorkspaceRoot(projectPath),
                0,
                getProjectNavigationScope(projectPath),
              ),
            },
          ],
        }),
        { tag: 'hr' },
        {
          tag: 'markdown',
          content: responseText,
          text_size: 'normal',
        },
      ],
    },
  });
}

async function sendFeishuNavigationCard(
  adapter: BaseChannelAdapter,
  address: ChannelAddress,
  cardJson: string,
  updateMessageId?: string,
): Promise<boolean> {
  if (adapter.channelType !== 'feishu') return false;
  const result = await deliver(adapter, {
    address,
    text: cardJson,
    parseMode: 'CardJson',
    updateMessageId,
  });
  if (!result.ok && updateMessageId) {
    console.warn('[bridge-manager] Feishu navigation card update failed without create fallback:', result.error || 'unknown error');
  }
  return result.ok;
}

async function sendReplyNavigationCard(
  adapter: BaseChannelAdapter,
  address: ChannelAddress,
  binding: ChannelBinding,
): Promise<boolean> {
  return sendFeishuNavigationCard(adapter, address, buildFeishuReplyNavCard(binding));
}

async function sendFeishuResponseCard(
  adapter: BaseChannelAdapter,
  address: ChannelAddress,
  binding: ChannelBinding,
  responseText: string,
): Promise<boolean> {
  if (adapter.channelType !== 'feishu') return false;
  if (!responseText || responseText.length > 12000) return false;
  return sendFeishuNavigationCard(adapter, address, buildFeishuResponseCard(binding, responseText));
}

function buildStatusFallbackText(binding: ChannelBinding): string {
  binding = syncBindingToCurrentSession(binding);
  return [
    '<b>Bridge Status</b>',
    '',
    `Project: <b>${escapeHtml(getProjectLabel(normalizeProjectPath(binding.workingDirectory)))}</b>`,
    `Path: <code>${escapeHtml(normalizeProjectPath(binding.workingDirectory))}</code>`,
    `Session: <code>${binding.codepilotSessionId.slice(0, 8)}...</code>`,
    `Mode: <b>${binding.mode}</b>`,
    `Model: <code>${escapeHtml(binding.model || 'default')}</code>`,
  ].join('\n');
}

function buildSessionsFallbackText(groups: ProjectGroup[]): string {
  if (groups.length === 0) {
    return 'No sessions found.';
  }

  const lines = ['<b>Projects</b>', ''];
  for (const group of groups.slice(0, 12)) {
    lines.push(`<b>${escapeHtml(group.label)}</b>`);
    lines.push(`<code>${escapeHtml(group.path)}</code>`);
    for (const session of group.sessions.slice(0, 5)) {
      lines.push(`- <code>${session.id.slice(0, 8)}...</code> ${escapeHtml(session.model || 'default')}`);
    }
    lines.push('');
  }
  return lines.join('\n').trim();
}

async function sendStatusCard(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
  binding: ChannelBinding,
  contextPage = 0,
  contextDetailIndex?: number,
): Promise<boolean> {
  return sendFeishuNavigationCard(adapter, msg.address, buildFeishuStatusCard(binding, contextPage, contextDetailIndex), msg.callbackMessageId);
}

async function sendProjectListCard(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
  binding: ChannelBinding,
  scope: ProjectListScope = 'global',
  workspacePath?: string | null,
): Promise<boolean> {
  const groups = buildProjectGroups(binding, scope, workspacePath);
  return sendFeishuNavigationCard(adapter, msg.address, buildFeishuProjectListCard(groups, binding, scope, workspacePath), msg.callbackMessageId);
}

function getProjectGroup(
  binding: ChannelBinding,
  projectPath: string,
  workspacePath?: string | null,
): ProjectGroup | undefined {
  const normalizedProjectPath = normalizeProjectPath(projectPath);
  return buildProjectGroups(binding, 'global', workspacePath).find((item) => item.path === normalizedProjectPath);
}

function parseProjectNavPayload(payload: string): {
  projectPath: string | null;
  scope: ProjectListScope;
  workspacePath: string | null;
  page: number;
} {
  const parts = payload.split(':');
  let scope: ProjectListScope = 'workspace';
  let encodedPath = payload;
  let workspacePath: string | null = null;
  let page = 0;

  if ((parts[0] === 'workspace' || parts[0] === 'global') && parts[1]) {
    scope = parts[0];
    encodedPath = parts[1];
    if (parts[2]) {
      page = Math.max(0, parseInt(parts[2], 10) || 0);
    }
  } else {
    encodedPath = parts[0] || '';
    workspacePath = parts[1] ? decodeCardToken(parts[1]) : null;
    scope = workspacePath ? 'global' : 'workspace';
    if (parts[2]) {
      page = Math.max(0, parseInt(parts[2], 10) || 0);
    }
  }

  return {
    projectPath: decodeCardToken(encodedPath),
    scope,
    workspacePath,
    page,
  };
}

async function sendProjectSessionsCard(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
  binding: ChannelBinding,
  projectPath: string,
  scope: ProjectListScope = 'workspace',
  workspacePath?: string | null,
  page = 0,
): Promise<boolean> {
  const group = getProjectGroup(binding, projectPath, workspacePath);
  if (!group) return false;
  return sendFeishuNavigationCard(
    adapter,
    msg.address,
    buildFeishuProjectSessionsCard(group, binding.codepilotSessionId, binding, scope, workspacePath, page),
    msg.callbackMessageId,
  );
}

async function sendSessionPreviewCard(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
  binding: ChannelBinding,
  sessionId: string,
  contextPage = 0,
  contextDetailIndex?: number,
): Promise<boolean> {
  const card = buildFeishuSessionPreviewCard(sessionId, binding.codepilotSessionId, contextPage, contextDetailIndex);
  if (!card) return false;
  return sendFeishuNavigationCard(adapter, msg.address, card, msg.callbackMessageId);
}

function archiveSessionAndRebindIfNeeded(
  address: ChannelAddress,
  binding: ChannelBinding,
  sessionId: string,
): ChannelBinding | null {
  const store = getSessionListingStore();
  const session = store.getSession(sessionId);
  if (!session) return null;

  store.archiveSession?.(sessionId);
  if (binding.codepilotSessionId !== sessionId) {
    return binding;
  }

  const replacement = store.createSession(
    `Bridge: ${address.displayName || address.chatId}`,
    session.model,
    session.system_prompt,
    session.working_directory,
    binding.mode,
  );
  if (session.provider_id) {
    store.updateSessionProviderId(replacement.id, session.provider_id);
  }
  return store.upsertChannelBinding({
    channelType: address.channelType,
    chatId: address.chatId,
    codepilotSessionId: replacement.id,
    sdkSessionId: '',
    workingDirectory: replacement.working_directory,
    model: replacement.model,
    mode: binding.mode,
  });
}

function ensureBindableSession(sessionId: string): NamedBridgeSession | null {
  const store = getSessionListingStore();
  const direct = store.getSession(sessionId) as NamedBridgeSession | null;
  if (direct) return direct;

  const mirrored = listBridgeSessions().find((session) => getStoredSdkSessionId(session) === sessionId);
  if (mirrored) return mirrored;

  const discovered = discoverCodexSessions().find((session) => session.id === sessionId);
  if (!discovered) return null;

  const name = discovered.name || `Session ${getShortSessionId(sessionId)}`;
  const created = store.createSession(
    name,
    discovered.model || store.getSetting('bridge_default_model') || '',
    discovered.system_prompt,
    discovered.working_directory,
    'code',
    {
      createdAt: discovered.created_at,
      updatedAt: discovered.updated_at,
    },
  ) as NamedBridgeSession;
  store.updateSessionName?.(created.id, name);
  store.updateSdkSessionId?.(created.id, sessionId);
  return (store.getSession(created.id) as NamedBridgeSession | null) ?? { ...created, sdk_session_id: sessionId };
}

async function handleNavigationCallback(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
): Promise<boolean> {
  const callbackData = msg.callbackData;
  if (!callbackData || !callbackData.startsWith('nav:')) {
    return false;
  }

  const binding = syncBindingToCurrentSession(router.resolve(msg.address));

  if (callbackData === 'nav:projects' || callbackData.startsWith('nav:projects:')) {
    const scopeToken = callbackData.split(':')[2];
    const scope: ProjectListScope = scopeToken === 'workspace' ? 'workspace' : 'global';
    const sent = await sendProjectListCard(adapter, msg, binding, scope);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildSessionsFallbackText(buildProjectGroups(binding, scope)),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }
  if (callbackData.startsWith('nav:workspace:')) {
    const workspacePath = decodeCardToken(callbackData.slice('nav:workspace:'.length));
    if (!workspacePath) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Workspace link is invalid or expired.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const sent = await sendProjectListCard(adapter, msg, binding, 'global', workspacePath);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildSessionsFallbackText(buildProjectGroups(binding, 'global', workspacePath)),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }
  if (callbackData.startsWith('nav:status')) {
    const parts = callbackData.split(':');
    const contextPage = Math.max(0, parseInt(parts[2] || '0', 10) || 0);
    const sent = await sendStatusCard(adapter, msg, binding, contextPage);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(binding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:ctx:status:')) {
    const parts = callbackData.split(':');
    const contextPage = Math.max(0, parseInt(parts[3] || '0', 10) || 0);
    const detailIndex = Math.max(0, parseInt(parts[4] || '0', 10) || 0);
    const sent = await sendStatusCard(adapter, msg, binding, contextPage, detailIndex);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(binding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:ctxclear:status:')) {
    const parts = callbackData.split(':');
    const contextPage = Math.max(0, parseInt(parts[3] || '0', 10) || 0);
    const sent = await sendStatusCard(adapter, msg, binding, contextPage);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(binding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:project:')) {
    const payload = callbackData.slice('nav:project:'.length);
    const {
      projectPath,
      scope,
      workspacePath,
      page,
    } = parseProjectNavPayload(payload);
    if (!projectPath) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Project link is invalid or expired.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const group = getProjectGroup(binding, projectPath, workspacePath);
    if (!group) {
      await deliver(adapter, {
        address: msg.address,
        text: `Project not found: <code>${escapeHtml(projectPath)}</code>`,
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const sent = await sendFeishuNavigationCard(
      adapter,
      msg.address,
      buildFeishuProjectSessionsCard(group, binding.codepilotSessionId, binding, scope, workspacePath, page),
      msg.callbackMessageId,
    );
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildSessionsFallbackText([group]),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:new:')) {
    const payload = callbackData.slice('nav:new:'.length);
    const {
      projectPath,
      scope,
      workspacePath,
    } = parseProjectNavPayload(payload);
    if (!projectPath) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Project link is invalid or expired.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }

    const nextBinding = router.createBinding(msg.address, projectPath);
    const dockBinding = markDockSessionSeen(addSessionToDock(nextBinding, nextBinding.codepilotSessionId, true), nextBinding.codepilotSessionId);
    const sent = await sendProjectSessionsCard(
      adapter,
      msg,
      dockBinding,
      projectPath,
      scope,
      workspacePath,
      0,
    );
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(dockBinding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:bind:')) {
    const sessionId = callbackData.slice('nav:bind:'.length);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const bindableSession = ensureBindableSession(sessionId);
    const boundSessionId = bindableSession?.id || sessionId;
    const nextBinding = bindableSession ? router.bindToSession(msg.address, boundSessionId) : null;
    if (!nextBinding) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    console.log('[bridge-manager] Session rebound:', JSON.stringify({
      chatId: msg.address.chatId,
      sessionId: boundSessionId,
      runtimeSessionId: getStoredSdkSessionId(bindableSession),
      workingDirectory: nextBinding.workingDirectory,
      model: nextBinding.model || null,
    }));
    const dockBinding = markDockSessionSeen(addSessionToDock(nextBinding, boundSessionId, true), boundSessionId);
    const sent = await sendStatusCard(adapter, msg, dockBinding);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(dockBinding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:dock:select:')) {
    const sessionId = callbackData.slice('nav:dock:select:'.length);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const bindableSession = ensureBindableSession(sessionId);
    const boundSessionId = bindableSession?.id || sessionId;
    const nextBinding = bindableSession ? router.bindToSession(msg.address, boundSessionId) : null;
    if (!nextBinding) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const dockBinding = markDockSessionSeen(addSessionToDock(nextBinding, boundSessionId, true), boundSessionId);
    const sent = await sendStatusCard(adapter, msg, dockBinding);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(dockBinding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:dock:close:')) {
    const sessionId = callbackData.slice('nav:dock:close:'.length);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const openIds = getOpenSessionIds(binding).filter((id) => id !== sessionId);
    if (binding.codepilotSessionId === sessionId && openIds.length > 0) {
      const nextBinding = router.bindToSession(msg.address, openIds[0]!);
      if (!nextBinding) {
        await deliver(adapter, {
          address: msg.address,
          text: 'Session not found.',
          parseMode: 'plain',
          replyToMessageId: msg.messageId,
        });
        return true;
      }
      const dockBinding = markDockSessionSeen(
        persistSessionDock(nextBinding, openIds, getDockSeenCounts(removeSessionFromDock(binding, sessionId))),
        nextBinding.codepilotSessionId,
      );
      const sent = await sendStatusCard(adapter, msg, dockBinding);
      if (!sent) {
        await deliver(adapter, {
          address: msg.address,
          text: buildStatusFallbackText(dockBinding),
          parseMode: 'HTML',
          replyToMessageId: msg.messageId,
        });
      }
      return true;
    }

    const dockBinding = removeSessionFromDock(binding, sessionId);
    const sent = await sendStatusCard(adapter, msg, dockBinding);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: buildStatusFallbackText(dockBinding),
        parseMode: 'HTML',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:peek:')) {
    const peekPayload = callbackData.slice('nav:peek:'.length);
    const [sessionId, pageToken] = peekPayload.split(':');
    const contextPage = Math.max(0, parseInt(pageToken || '0', 10) || 0);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const sent = await sendSessionPreviewCard(adapter, msg, binding, sessionId, contextPage);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:ctx:peek:')) {
    const peekPayload = callbackData.slice('nav:ctx:peek:'.length);
    const [sessionId, pageToken, detailToken] = peekPayload.split(':');
    const contextPage = Math.max(0, parseInt(pageToken || '0', 10) || 0);
    const detailIndex = Math.max(0, parseInt(detailToken || '0', 10) || 0);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const sent = await sendSessionPreviewCard(adapter, msg, binding, sessionId, contextPage, detailIndex);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:ctxclear:peek:')) {
    const peekPayload = callbackData.slice('nav:ctxclear:peek:'.length);
    const [sessionId, pageToken] = peekPayload.split(':');
    const contextPage = Math.max(0, parseInt(pageToken || '0', 10) || 0);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const sent = await sendSessionPreviewCard(adapter, msg, binding, sessionId, contextPage);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  if (callbackData.startsWith('nav:archive:')) {
    const sessionId = callbackData.slice('nav:archive:'.length);
    if (!validateSessionId(sessionId)) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Invalid session ID.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const nextBinding = archiveSessionAndRebindIfNeeded(msg.address, binding, sessionId);
    if (!nextBinding) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session not found.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
      return true;
    }
    const archivedSession = getSessionListingStore().getSession(sessionId);
    const projectPath = normalizeProjectPath(archivedSession?.working_directory || nextBinding.workingDirectory);
    const sent = await sendProjectSessionsCard(adapter, msg, nextBinding, projectPath);
    if (!sent) {
      await deliver(adapter, {
        address: msg.address,
        text: 'Session archived.',
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    }
    return true;
  }

  return false;
}

interface AdapterMeta {
  lastMessageAt: string | null;
  lastError: string | null;
}

interface BridgeManagerState {
  adapters: Map<string, BaseChannelAdapter>;
  adapterMeta: Map<string, AdapterMeta>;
  running: boolean;
  startedAt: string | null;
  loopAborts: Map<string, AbortController>;
  activeTasks: Map<string, AbortController>;
  /** Per-session processing chains for concurrency control */
  sessionLocks: Map<string, Promise<void>>;
  autoStartChecked: boolean;
}

function getState(): BridgeManagerState {
  const g = globalThis as unknown as Record<string, BridgeManagerState>;
  if (!g[GLOBAL_KEY]) {
    g[GLOBAL_KEY] = {
      adapters: new Map(),
      adapterMeta: new Map(),
      running: false,
      startedAt: null,
      loopAborts: new Map(),
      activeTasks: new Map(),
      sessionLocks: new Map(),
      autoStartChecked: false,
    };
  }
  // Backfill sessionLocks for states created before this field existed
  if (!g[GLOBAL_KEY].sessionLocks) {
    g[GLOBAL_KEY].sessionLocks = new Map();
  }
  return g[GLOBAL_KEY];
}

/**
 * Process a function with per-session serialization.
 * Different sessions run concurrently; same-session requests are serialized.
 */
function processWithSessionLock(sessionId: string, fn: () => Promise<void>): Promise<void> {
  const state = getState();
  const prev = state.sessionLocks.get(sessionId) || Promise.resolve();
  const current = prev.then(fn, fn);
  state.sessionLocks.set(sessionId, current);
  // Cleanup when the chain completes.
  // Suppress rejection on the cleanup chain 閳?callers handle errors on `current` directly.
  current.finally(() => {
    if (state.sessionLocks.get(sessionId) === current) {
      state.sessionLocks.delete(sessionId);
    }
  }).catch(() => {});
  return current;
}

/**
 * Start the bridge system.
 * Checks feature flags, registers enabled adapters, starts polling loops.
 */
export async function start(): Promise<void> {
  const state = getState();
  if (state.running) return;

  const { store, lifecycle } = getBridgeContext();

  const bridgeEnabled = store.getSetting('remote_bridge_enabled') === 'true';
  if (!bridgeEnabled) {
    console.log('[bridge-manager] Bridge not enabled (remote_bridge_enabled != true)');
    return;
  }

  // Iterate all registered adapter types and create those that are enabled
  for (const channelType of getRegisteredTypes()) {
    const settingKey = `bridge_${channelType}_enabled`;
    if (store.getSetting(settingKey) !== 'true') continue;

    const adapter = createAdapter(channelType);
    if (!adapter) continue;

    const configError = adapter.validateConfig();
    if (!configError) {
      registerAdapter(adapter);
    } else {
      console.warn(`[bridge-manager] ${channelType} adapter not valid:`, configError);
    }
  }

  // Start all registered adapters, track how many succeeded
  let startedCount = 0;
  for (const [type, adapter] of state.adapters) {
    try {
      await adapter.start();
      console.log(`[bridge-manager] Started adapter: ${type}`);
      startedCount++;
    } catch (err) {
      console.error(`[bridge-manager] Failed to start adapter ${type}:`, err);
    }
  }

  // Only mark as running if at least one adapter started successfully
  if (startedCount === 0) {
    console.warn('[bridge-manager] No adapters started successfully, bridge not activated');
    state.adapters.clear();
    state.adapterMeta.clear();
    return;
  }

  // Mark running BEFORE starting consumer loops 閳?runAdapterLoop checks
  // state.running in its while-condition, so it must be true first.
  state.running = true;
  state.startedAt = new Date().toISOString();

  // Notify host that bridge is starting (e.g., suppress competing polling)
  lifecycle.onBridgeStart?.();

  // Now start the consumer loops (state.running is already true)
  for (const [, adapter] of state.adapters) {
    if (adapter.isRunning()) {
      runAdapterLoop(adapter);
    }
  }

  console.log(`[bridge-manager] Bridge started with ${startedCount} adapter(s)`);
}

/**
 * Stop the bridge system gracefully.
 */
export async function stop(): Promise<void> {
  const state = getState();
  if (!state.running) return;

  const { lifecycle } = getBridgeContext();

  state.running = false;

  // Abort all event loops
  for (const [, abort] of state.loopAborts) {
    abort.abort();
  }
  state.loopAborts.clear();

  // Stop all adapters
  for (const [type, adapter] of state.adapters) {
    try {
      await adapter.stop();
      console.log(`[bridge-manager] Stopped adapter: ${type}`);
    } catch (err) {
      console.error(`[bridge-manager] Error stopping adapter ${type}:`, err);
    }
  }

  state.adapters.clear();
  state.adapterMeta.clear();
  state.startedAt = null;

  // Notify host that bridge stopped
  lifecycle.onBridgeStop?.();

  console.log('[bridge-manager] Bridge stopped');
}

/**
 * Lazy auto-start: checks bridge_auto_start setting once and starts if enabled.
 * Called from POST /api/bridge with action 'auto-start' (triggered by Electron on startup).
 */
export function tryAutoStart(): void {
  const state = getState();
  if (state.autoStartChecked) return;
  state.autoStartChecked = true;

  if (state.running) return;

  const { store } = getBridgeContext();
  const autoStart = store.getSetting('bridge_auto_start');
  if (autoStart !== 'true') return;

  start().catch(err => {
    console.error('[bridge-manager] Auto-start failed:', err);
  });
}

/**
 * Get the current bridge status.
 */
export function getStatus(): BridgeStatus {
  const state = getState();
  return {
    running: state.running,
    startedAt: state.startedAt,
    adapters: Array.from(state.adapters.entries()).map(([type, adapter]) => {
      const meta = state.adapterMeta.get(type);
      return {
        channelType: adapter.channelType,
        running: adapter.isRunning(),
        connectedAt: state.startedAt,
        lastMessageAt: meta?.lastMessageAt ?? null,
        error: meta?.lastError ?? null,
      };
    }),
  };
}

/**
 * Register a channel adapter.
 */
export function registerAdapter(adapter: BaseChannelAdapter): void {
  const state = getState();
  state.adapters.set(adapter.channelType, adapter);
}

/**
 * Run the event loop for a single adapter.
 * Messages for different sessions are dispatched concurrently;
 * messages for the same session are serialized via session locks.
 */
function runAdapterLoop(adapter: BaseChannelAdapter): void {
  const state = getState();
  const abort = new AbortController();
  state.loopAborts.set(adapter.channelType, abort);

  (async () => {
    while (state.running && adapter.isRunning()) {
      try {
        const msg = await adapter.consumeOne();
        if (!msg) continue; // Adapter stopped

        // Callback queries, commands, and numeric permission shortcuts are
        // lightweight 閳?process inline (outside session lock).
        // Regular messages use per-session locking for concurrency.
        //
        // IMPORTANT: numeric shortcuts (1/2/3) for feishu/qq MUST run outside
        // the session lock. The current session is blocked waiting for the
        // permission to be resolved; if "1" enters the session lock queue it
        // deadlocks (permission waits for "1", "1" waits for lock release).
        if (
          msg.callbackData ||
          msg.text.trim().startsWith('/') ||
          isNumericPermissionShortcut(adapter.channelType, msg.text.trim(), msg.address.chatId)
        ) {
          await handleMessage(adapter, msg);
        } else {
          const binding = syncBindingToCurrentSession(router.resolve(msg.address));
          // Fire-and-forget into session lock 閳?loop continues to accept
          // messages for other sessions immediately.
          processWithSessionLock(binding.codepilotSessionId, () =>
            handleMessage(adapter, msg),
          ).catch(err => {
            console.error(`[bridge-manager] Session ${binding.codepilotSessionId.slice(0, 8)} error:`, err);
          });
        }
      } catch (err) {
        if (abort.signal.aborted) break;
        const errMsg = err instanceof Error ? err.message : String(err);
        console.error(`[bridge-manager] Error in ${adapter.channelType} loop:`, err);
        // Track last error per adapter
        const meta = state.adapterMeta.get(adapter.channelType) || { lastMessageAt: null, lastError: null };
        meta.lastError = errMsg;
        state.adapterMeta.set(adapter.channelType, meta);
        // Brief delay to prevent tight error loops
        await new Promise(r => setTimeout(r, 1000));
      }
    }
  })().catch(err => {
    if (!abort.signal.aborted) {
      const errMsg = err instanceof Error ? err.message : String(err);
      console.error(`[bridge-manager] ${adapter.channelType} loop crashed:`, err);
      const meta = state.adapterMeta.get(adapter.channelType) || { lastMessageAt: null, lastError: null };
      meta.lastError = errMsg;
      state.adapterMeta.set(adapter.channelType, meta);
    }
  });
}

/**
 * Handle a single inbound message.
 */
async function handleMessage(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
): Promise<void> {
  const { store } = getBridgeContext();

  // Update lastMessageAt for this adapter
  const adapterState = getState();
  const meta = adapterState.adapterMeta.get(adapter.channelType) || { lastMessageAt: null, lastError: null };
  meta.lastMessageAt = new Date().toISOString();
  adapterState.adapterMeta.set(adapter.channelType, meta);

  // Acknowledge the update offset after processing completes (or fails).
  // This ensures the adapter only advances its committed offset once the
  // message has been fully handled, preventing message loss on crash.
  const ack = () => {
    if (msg.updateId != null && adapter.acknowledgeUpdate) {
      adapter.acknowledgeUpdate(msg.updateId);
    }
  };

  // Handle callback queries (permission buttons)
  if (msg.callbackData) {
    const navigationHandled = await handleNavigationCallback(adapter, msg);
    if (navigationHandled) {
      ack();
      return;
    }

    const handled = broker.handlePermissionCallback(msg.callbackData, msg.address.chatId, msg.callbackMessageId);
    if (handled) {
      // Send confirmation
      const confirmMsg: OutboundMessage = {
        address: msg.address,
        text: 'Permission response recorded.',
        parseMode: 'plain',
      };
      await deliver(adapter, confirmMsg);
    }
    ack();
    return;
  }

  const rawText = msg.text.trim();
  const hasAttachments = msg.attachments && msg.attachments.length > 0;

  // Handle attachment-only download failures 閳?surface error to user instead of silently dropping
  if (!rawText && !hasAttachments) {
    const rawData = msg.raw as {
      imageDownloadFailed?: boolean;
      attachmentDownloadFailed?: boolean;
      failedCount?: number;
      failedLabel?: string;
      userVisibleError?: string;
    } | undefined;
    if (rawData?.userVisibleError) {
      await deliver(adapter, {
        address: msg.address,
        text: rawData.userVisibleError,
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    } else if (rawData?.imageDownloadFailed || rawData?.attachmentDownloadFailed) {
      const failureLabel = rawData.failedLabel || (rawData.imageDownloadFailed ? 'image(s)' : 'attachment(s)');
      await deliver(adapter, {
        address: msg.address,
        text: `Failed to download ${rawData.failedCount ?? 1} ${failureLabel}. Please try sending again.`,
        parseMode: 'plain',
        replyToMessageId: msg.messageId,
      });
    }
    ack();
    return;
  }

  // 閳光偓閳光偓 Numeric shortcut for permission replies (feishu/qq/weixin only) 閳光偓閳光偓
  // On mobile, typing `/perm allow <uuid>` is painful.
  // If the user sends "1", "2", or "3" and there is exactly one pending
  // permission for this chat, map it: 1閳妺llow, 2閳妺llow_session, 3閳妿eny.
  //
  // Input normalization: mobile keyboards / IM clients may send fullwidth
  // digits (閿涙埊绱ラ敍?, digits with zero-width joiners, or other Unicode
  // variants. NFKC normalization folds them all to ASCII 1/2/3.
  if (
    adapter.channelType === 'feishu'
    || adapter.channelType === 'qq'
    || adapter.channelType === 'weixin'
  ) {
    // eslint-disable-next-line no-control-regex
    const normalized = rawText.normalize('NFKC').replace(/[\u200B-\u200D\uFEFF]/g, '').trim();
    if (/^[123]$/.test(normalized)) {
      const pendingLinks = store.listPendingPermissionLinksByChat(msg.address.chatId);
      if (pendingLinks.length === 1) {
        const actionMap: Record<string, string> = { '1': 'allow', '2': 'allow_session', '3': 'deny' };
        const action = actionMap[normalized];
        const permId = pendingLinks[0].permissionRequestId;
        const callbackData = `perm:${action}:${permId}`;
        const handled = broker.handlePermissionCallback(callbackData, msg.address.chatId);
        const label = normalized === '1' ? 'Allow' : normalized === '2' ? 'Allow Session' : 'Deny';
        if (handled) {
          await deliver(adapter, {
            address: msg.address,
            text: `${label}: recorded.`,
            parseMode: 'plain',
            replyToMessageId: msg.messageId,
          });
        } else {
          await deliver(adapter, {
            address: msg.address,
            text: `Permission not found or already resolved.`,
            parseMode: 'plain',
            replyToMessageId: msg.messageId,
          });
        }
        ack();
        return;
      }
      if (pendingLinks.length > 1) {
        // Multiple pending permissions 閳?numeric shortcut is ambiguous.
        await deliver(adapter, {
          address: msg.address,
          text: `Multiple pending permissions (${pendingLinks.length}). Please use the full command:\n/perm allow|allow_session|deny <id>`,
          parseMode: 'plain',
          replyToMessageId: msg.messageId,
        });
        ack();
        return;
      }
      // pendingLinks.length === 0: no pending permissions, fall through as normal message
    } else if (rawText !== normalized && /^[123]$/.test(rawText) === false) {
      // Log when normalization changed the text 閳?helps diagnose encoding issues
      const codePoints = [...rawText].map(c => 'U+' + c.codePointAt(0)!.toString(16).toUpperCase().padStart(4, '0'));
      console.log(`[bridge-manager] Shortcut candidate raw codepoints: ${codePoints.join(' ')} 閳?normalized: "${normalized}"`);
    }
  }

  // Check for IM commands (before sanitization 閳?commands are validated individually)
  if (rawText.startsWith('/')) {
    await handleCommand(adapter, msg, rawText);
    ack();
    return;
  }

  // Sanitize general message text before routing to conversation engine
  const { text, truncated } = sanitizeInput(rawText);
  if (truncated) {
    console.warn(`[bridge-manager] Input truncated from ${rawText.length} to ${text.length} chars for chat ${msg.address.chatId}`);
    store.insertAuditLog({
      channelType: adapter.channelType,
      chatId: msg.address.chatId,
      direction: 'inbound',
      messageId: msg.messageId,
      summary: `[TRUNCATED] Input truncated from ${rawText.length} chars`,
    });
  }

  if (!text && !hasAttachments) { ack(); return; }

  // Regular message 閳?route to conversation engine
  const binding = syncBindingToCurrentSession(router.resolve(msg.address));

  // Notify adapter that message processing is starting (e.g., typing indicator)
  adapter.onMessageStart?.(msg.address.chatId);

  // Create an AbortController so /stop can cancel this task externally
  const taskAbort = new AbortController();
  const state = getState();
  state.activeTasks.set(binding.codepilotSessionId, taskAbort);

  // 閳光偓閳光偓 Streaming preview setup 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓
  let previewState: StreamingPreviewState | null = null;
  const caps = adapter.getPreviewCapabilities?.(msg.address.chatId) ?? null;
  if (caps?.supported) {
    previewState = {
      draftId: generateDraftId(),
      chatId: msg.address.chatId,
      lastSentText: '',
      lastSentAt: 0,
      degraded: false,
      throttleTimer: null,
      pendingText: '',
    };
  }

  const streamCfg = previewState ? getStreamConfig(adapter.channelType) : null;

  // Build the preview onPartialText callback (or undefined if preview not supported)
  const previewOnPartialText = (previewState && streamCfg) ? (fullText: string) => {
    const ps = previewState!;
    const cfg = streamCfg!;
    if (ps.degraded) return;

    // Truncate to maxChars + ellipsis
    ps.pendingText = fullText.length > cfg.maxChars
      ? fullText.slice(0, cfg.maxChars) + '...'
      : fullText;

    const delta = ps.pendingText.length - ps.lastSentText.length;
    const elapsed = Date.now() - ps.lastSentAt;

    if (delta < cfg.minDeltaChars && ps.lastSentAt > 0) {
      // Not enough new content 閳?schedule trailing-edge timer if not already set
      if (!ps.throttleTimer) {
        ps.throttleTimer = setTimeout(() => {
          ps.throttleTimer = null;
          if (!ps.degraded) flushPreview(adapter, ps, cfg);
        }, cfg.intervalMs);
      }
      return;
    }

    if (elapsed < cfg.intervalMs && ps.lastSentAt > 0) {
      // Too soon 閳?schedule trailing-edge timer to ensure latest text is sent
      if (!ps.throttleTimer) {
        ps.throttleTimer = setTimeout(() => {
          ps.throttleTimer = null;
          if (!ps.degraded) flushPreview(adapter, ps, cfg);
        }, cfg.intervalMs - elapsed);
      }
      return;
    }

    // Clear any pending trailing-edge timer and flush immediately
    if (ps.throttleTimer) {
      clearTimeout(ps.throttleTimer);
      ps.throttleTimer = null;
    }
    flushPreview(adapter, ps, cfg);
  } : undefined;

  // 閳光偓閳光偓 Streaming card setup (Feishu CardKit v2) 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓
  // If the adapter supports streaming cards (e.g. Feishu), wire up
  // onStreamText, onToolEvent, and onStreamEnd callbacks.
  // These run in parallel with the existing preview system 閳?Feishu
  // uses cards instead of message edit for streaming.
  const hasStreamingCards = typeof adapter.onStreamText === 'function';
  const toolCallTracker = new Map<string, ToolCallInfo>();

  const onStreamCardText = hasStreamingCards ? (fullText: string) => {
    try { adapter.onStreamText!(msg.address.chatId, fullText); } catch { /* non-critical */ }
  } : undefined;

  const onToolEvent = hasStreamingCards ? (toolId: string, toolName: string, status: 'running' | 'complete' | 'error') => {
    if (toolName) {
      toolCallTracker.set(toolId, { id: toolId, name: toolName, status });
    } else {
      // tool_result doesn't carry name 閳?update existing entry's status
      const existing = toolCallTracker.get(toolId);
      if (existing) existing.status = status;
    }
    try {
      adapter.onToolEvent!(msg.address.chatId, Array.from(toolCallTracker.values()));
    } catch { /* non-critical */ }
  } : undefined;

  // Combined partial text callback: streaming preview + streaming cards
  const onPartialText = (previewOnPartialText || onStreamCardText) ? (fullText: string) => {
    if (previewOnPartialText) previewOnPartialText(fullText);
    if (onStreamCardText) onStreamCardText(fullText);
  } : undefined;

  try {
    // Pass permission callback so requests are forwarded to IM immediately
    // during streaming (the stream blocks until permission is resolved).
    // Use text or empty string for image-only messages (prompt is still required by streamClaude)
    const promptText = text || (hasAttachments ? 'Describe this image.' : '');

    const result = await engine.processMessage(binding, promptText, async (perm) => {
      await broker.forwardPermissionRequest(
        adapter,
        msg.address,
        perm.permissionRequestId,
        perm.toolName,
        perm.toolInput,
        binding.codepilotSessionId,
        perm.suggestions,
        msg.messageId,
      );
    }, taskAbort.signal, hasAttachments ? msg.attachments : undefined, onPartialText, onToolEvent);

    // Finalize streaming card if adapter supports it.
    // onStreamEnd awaits any in-flight card creation and returns true if a card
    // was actually finalized (meaning content is already visible to the user).
    let cardFinalized = false;
    if (hasStreamingCards && adapter.onStreamEnd) {
      try {
        const status = result.hasError ? 'error' : 'completed';
        cardFinalized = await adapter.onStreamEnd(msg.address.chatId, status, result.responseText);
      } catch (err) {
        console.warn('[bridge-manager] Card finalize failed:', err instanceof Error ? err.message : err);
      }
    }

    // Send response text 閳?render via channel-appropriate format.
    // Skip if streaming card was finalized (content already in card).
    if (result.responseText) {
      const sentAsHeaderCard = !cardFinalized
        ? await sendFeishuResponseCard(adapter, msg.address, binding, result.responseText)
        : false;
      if (!cardFinalized && !sentAsHeaderCard) {
        await deliverResponse(adapter, msg.address, result.responseText, binding.codepilotSessionId, msg.messageId);
      }
    } else if (result.hasError) {
      const sentErrorCard = !cardFinalized
        ? await sendFeishuResponseCard(
            adapter,
            msg.address,
            binding,
            `**Error**\n\n${result.errorMessage}`,
          )
        : false;
      if (!cardFinalized && !sentErrorCard) {
        const errorResponse: OutboundMessage = {
          address: msg.address,
          text: `<b>Error:</b> ${escapeHtml(result.errorMessage)}`,
          parseMode: 'HTML',
          replyToMessageId: msg.messageId,
        };
        await deliver(adapter, errorResponse);
      }
    }

    // Persist the actual SDK session ID for future resume.
    // If the result has an error and no session ID was captured, clear the
    // stale ID so the next message starts fresh instead of retrying a broken resume.
    if (binding.id) {
      try {
        const update = computeSdkSessionUpdate(result.sdkSessionId, result.hasError);
        if (update !== null) {
          store.updateChannelBinding(binding.id, { sdkSessionId: update });
        }
      } catch { /* best effort */ }
    }
  } finally {
    // Clean up preview state
    if (previewState) {
      if (previewState.throttleTimer) {
        clearTimeout(previewState.throttleTimer);
        previewState.throttleTimer = null;
      }
      adapter.endPreview?.(msg.address.chatId, previewState.draftId);
    }

    // If task was aborted and streaming card is still active, finalize as interrupted
    if (hasStreamingCards && adapter.onStreamEnd && taskAbort.signal.aborted) {
      try {
        await adapter.onStreamEnd(msg.address.chatId, 'interrupted', '');
      } catch { /* best effort */ }
    }

    state.activeTasks.delete(binding.codepilotSessionId);
    // Notify adapter that message processing ended
    adapter.onMessageEnd?.(msg.address.chatId);
    // Commit the offset only after full processing (success or failure)
    ack();
  }
}

/**
 * Handle IM slash commands.
 */
async function handleCommand(
  adapter: BaseChannelAdapter,
  msg: InboundMessage,
  text: string,
): Promise<void> {
  const { store } = getBridgeContext();

  // Extract command and args (handle /command@botname format)
  const parts = text.split(/\s+/);
  const command = parts[0].split('@')[0].toLowerCase();
  const args = parts.slice(1).join(' ').trim();

  // Run dangerous-input detection on the full command text
  const dangerCheck = isDangerousInput(text);
  if (dangerCheck.dangerous) {
    store.insertAuditLog({
      channelType: adapter.channelType,
      chatId: msg.address.chatId,
      direction: 'inbound',
      messageId: msg.messageId,
      summary: `[BLOCKED] Dangerous input detected: ${dangerCheck.reason}`,
    });
    console.warn(`[bridge-manager] Blocked dangerous command input from chat ${msg.address.chatId}: ${dangerCheck.reason}`);
    await deliver(adapter, {
      address: msg.address,
      text: `Command rejected: invalid input detected.`,
      parseMode: 'plain',
      replyToMessageId: msg.messageId,
    });
    return;
  }

  let response = '';

  switch (command) {
    case '/start':
      response = [
        '<b>CodePilot Bridge</b>',
        '',
        'Send any message to interact with Claude.',
        '',
        '<b>Commands:</b>',
        '/new [path] - Start new session',
        '/bind &lt;session_id&gt; - Bind to existing session',
        '/cwd /path - Change working directory',
        '/mode plan|code|ask - Change mode',
        '/status - Show current status',
        '/sessions - List recent sessions',
        '/stop - Stop current session',
        '/perm allow|allow_session|deny &lt;id&gt; - Respond to permission',
        '/help - Show this help',
      ].join('\n');
      break;

    case '/new': {
      let workDir: string | undefined;
      if (args) {
        const validated = validateWorkingDirectory(args);
        if (!validated) {
          response = 'Invalid path. Must be an absolute path without traversal sequences.';
          break;
        }
        workDir = validated;
      }
      const binding = router.createBinding(msg.address, workDir);
      const dockBinding = markDockSessionSeen(addSessionToDock(binding, binding.codepilotSessionId, true), binding.codepilotSessionId);
      if (await sendStatusCard(adapter, msg, dockBinding)) {
        break;
      }
      response = `New session created.\nSession: <code>${dockBinding.codepilotSessionId.slice(0, 8)}...</code>\nCWD: <code>${escapeHtml(dockBinding.workingDirectory || '~')}</code>`;
      break;
    }

    case '/bind': {
      if (!args) {
        response = 'Usage: /bind &lt;session_id&gt;';
        break;
      }
      if (!validateSessionId(args)) {
        response = 'Invalid session ID format. Expected a 32-64 character hex/UUID string.';
        break;
      }
      const bindableSession = ensureBindableSession(args);
      const boundSessionId = bindableSession?.id || args;
      const binding = bindableSession ? router.bindToSession(msg.address, boundSessionId) : null;
      if (binding) {
        const dockBinding = markDockSessionSeen(addSessionToDock(binding, boundSessionId, true), boundSessionId);
        if (await sendStatusCard(adapter, msg, dockBinding)) {
          break;
        }
        response = `Bound to session <code>${boundSessionId.slice(0, 8)}...</code>`;
      } else {
        response = 'Session not found.';
      }
      break;
    }

    case '/cwd': {
      if (!args) {
        response = 'Usage: /cwd /path/to/directory';
        break;
      }
      const validatedPath = validateWorkingDirectory(args);
      if (!validatedPath) {
        response = 'Invalid path. Must be an absolute path without traversal sequences or special characters.';
        break;
      }
      const binding = router.resolve(msg.address);
      router.updateBinding(binding.id, { workingDirectory: validatedPath });
      response = `Working directory set to <code>${escapeHtml(validatedPath)}</code>`;
      break;
    }

    case '/mode': {
      if (!validateMode(args)) {
        response = 'Usage: /mode plan|code|ask';
        break;
      }
      const binding = router.resolve(msg.address);
      router.updateBinding(binding.id, { mode: args });
      response = `Mode set to <b>${args}</b>`;
      break;
    }

    case '/status': {
      const binding = router.resolve(msg.address);
      if (await sendStatusCard(adapter, msg, binding)) {
        break;
      }
      response = [
        '<b>Bridge Status</b>',
        '',
        `Session: <code>${binding.codepilotSessionId.slice(0, 8)}...</code>`,
        `CWD: <code>${escapeHtml(binding.workingDirectory || '~')}</code>`,
        `Mode: <b>${binding.mode}</b>`,
        `Model: <code>${binding.model || 'default'}</code>`,
      ].join('\n');
      break;
    }

    case '/sessions': {
      const binding = router.resolve(msg.address);
      const groups = buildProjectGroups(binding);
      if (await sendProjectListCard(adapter, msg, binding)) {
        break;
      }
      response = buildSessionsFallbackText(groups);
      break;
    }

    case '/stop': {
      const binding = router.resolve(msg.address);
      const st = getState();
      const taskAbort = st.activeTasks.get(binding.codepilotSessionId);
      if (taskAbort) {
        taskAbort.abort();
        st.activeTasks.delete(binding.codepilotSessionId);
        response = 'Stopping current task...';
      } else {
        response = 'No task is currently running.';
      }
      break;
    }

    case '/perm': {
      // Text-based permission approval fallback (for channels without inline buttons)
      // Usage: /perm allow <id> | /perm allow_session <id> | /perm deny <id>
      const permParts = args.split(/\s+/);
      const permAction = permParts[0];
      const permId = permParts.slice(1).join(' ');
      if (!permAction || !permId || !['allow', 'allow_session', 'deny'].includes(permAction)) {
        response = 'Usage: /perm allow|allow_session|deny &lt;permission_id&gt;';
        break;
      }
      const callbackData = `perm:${permAction}:${permId}`;
      const handled = broker.handlePermissionCallback(callbackData, msg.address.chatId);
      if (handled) {
        response = `Permission ${permAction}: recorded.`;
      } else {
        response = `Permission not found or already resolved.`;
      }
      break;
    }

    case '/help':
      response = [
        '<b>CodePilot Bridge Commands</b>',
        '',
        '/new [path] - Start new session',
        '/bind &lt;session_id&gt; - Bind to existing session',
        '/cwd /path - Change working directory',
        '/mode plan|code|ask - Change mode',
        '/status - Show current status',
        '/sessions - List recent sessions',
        '/stop - Stop current session',
        '/perm allow|allow_session|deny &lt;id&gt; - Respond to permission request',
        '1/2/3 - Quick permission reply (Feishu/QQ/WeChat, single pending)',
        '/help - Show this help',
      ].join('\n');
      break;

    default:
      response = `Unknown command: ${escapeHtml(command)}\nType /help for available commands.`;
  }

  if (response) {
    await deliver(adapter, {
      address: msg.address,
      text: response,
      parseMode: 'HTML',
      replyToMessageId: msg.messageId,
    });
  }
}

// 閳光偓閳光偓 SDK Session Update Logic 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓

/**
 * Compute the sdkSessionId value to persist after a conversation result.
 * Returns the new value to write, or null if no update is needed.
 *
 * Rules:
 * - If result has sdkSessionId AND no error 閳?save the new ID
 * - If result has error (regardless of sdkSessionId) 閳?clear to empty string
 * - Otherwise 閳?no update needed
 */
export function computeSdkSessionUpdate(
  sdkSessionId: string | null | undefined,
  hasError: boolean,
): string | null {
  if (sdkSessionId && !hasError) {
    return sdkSessionId;
  }
  if (hasError) {
    return '';
  }
  return null;
}

// 閳光偓閳光偓 Test-only export 閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓閳光偓
// Exposed so integration tests can exercise handleMessage directly
// without wiring up the full adapter loop.
/** @internal */
export const _testOnly = {
  buildProjectGroups,
  buildWorkspaceGroups,
  buildFeishuResponseCard,
  buildFeishuProjectListCard,
  buildFeishuReplyNavCard,
  buildFeishuProjectSessionsCard,
  buildFeishuSessionPreviewCard,
  buildFeishuStatusCard,
  clearCodexDiscoveryCache,
  decodeCardToken,
  encodeCardToken,
  getProjectLabel,
  sendFeishuNavigationCard,
  handleMessage,
};
